{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 01 - Data Exploration and labelling\n",
    "\n",
    "In Maffei.xlsx, in the sheet summary statistics, it is reported which phage causes lysis in which E.coli strain. Each row represents a phage, and the columns represent the host. The classification is binary (1=lysis, 0=no-lysis). For the sake of the project, we will consider successful infection only when there is lysis."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Import user defined libraries\n",
    "import sys\n",
    "sys.path.append(\"../src\")\n",
    "from fastaLoad import fasta"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.Load data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Maffei\n",
    "Let's load data from Maffei. The excel sheet is not clean for import and some rows and columns need to be skipped, as well as some header renamed.\\\n",
    "Also, we only care about the phage identifiers and lysis data, so we'll only keep those columns.\\\n",
    "We obtain a maffei dataset with the first 4 columns identifying the virus: bas, family, genus and phage.\\\n",
    "The following columns state host strains and contain binary values for lysis or not: 1 = lysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bas</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>phage</th>\n",
       "      <th>E_coli_UTI89</th>\n",
       "      <th>E_coli_CFT073</th>\n",
       "      <th>E_coli_55989</th>\n",
       "      <th>S_e_Typhimurium_12023s</th>\n",
       "      <th>S_e_Typhimurium_SL1344</th>\n",
       "      <th>E_coli_B_REL606</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bas01</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage AugustePiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bas02</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage JeanPiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bas03</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage JulesPiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bas04</td>\n",
       "      <td>Drexlerviridae; Tempevirinae</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage FritzSarasin</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bas05</td>\n",
       "      <td>Drexlerviridae; Tempevirinae</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage PeterMerian</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bas                        family         genus  \\\n",
       "0  Bas01  Drexlerviridae; Braunvirinae   Guelphvirus   \n",
       "1  Bas02  Drexlerviridae; Braunvirinae      Rtpvirus   \n",
       "2  Bas03  Drexlerviridae; Braunvirinae      Rtpvirus   \n",
       "3  Bas04  Drexlerviridae; Tempevirinae  Warwickvirus   \n",
       "4  Bas05  Drexlerviridae; Tempevirinae  Warwickvirus   \n",
       "\n",
       "                              phage  E_coli_UTI89  E_coli_CFT073  \\\n",
       "0  Escherichia phage AugustePiccard             0              0   \n",
       "1     Escherichia phage JeanPiccard             0              0   \n",
       "2    Escherichia phage JulesPiccard             0              0   \n",
       "3    Escherichia phage FritzSarasin             0              1   \n",
       "4     Escherichia phage PeterMerian             0              0   \n",
       "\n",
       "   E_coli_55989  S_e_Typhimurium_12023s  S_e_Typhimurium_SL1344  \\\n",
       "0             0                       0                       0   \n",
       "1             0                       0                       0   \n",
       "2             0                       0                       0   \n",
       "3             0                       0                       0   \n",
       "4             0                       0                       0   \n",
       "\n",
       "   E_coli_B_REL606  \n",
       "0                1  \n",
       "1                1  \n",
       "2                1  \n",
       "3                1  \n",
       "4                1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "bas                       string[python]\n",
       "family                    string[python]\n",
       "genus                     string[python]\n",
       "phage                     string[python]\n",
       "E_coli_UTI89                       int64\n",
       "E_coli_CFT073                      int64\n",
       "E_coli_55989                       int64\n",
       "S_e_Typhimurium_12023s             int64\n",
       "S_e_Typhimurium_SL1344             int64\n",
       "E_coli_B_REL606                    int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Maffei raw data\n",
    "RAW_PATH = os.path.join('..', 'data', 'raw')\n",
    "\n",
    "maffei = pd.read_excel(os.path.join(RAW_PATH, 'Maffei.xlsx'), sheet_name='summary statistics', skiprows=1, header=[0,1,2])\n",
    "maffei.drop(columns=maffei.columns[0], inplace=True)\n",
    "\n",
    "# Store the first level of the first 4 headers in a list\n",
    "header = [col[0] for col in maffei.columns.tolist()[:4]]\n",
    "header.append('lysis observed (yes=1, no=0)')\n",
    "\n",
    "# Keep only selected headers\n",
    "maffei = maffei.loc[:, header]\n",
    "\n",
    "# Drop odd columns up to 7\n",
    "maffei.drop(columns=maffei.columns[1:9:2], inplace=True)\n",
    "\n",
    "# Remove the first two rows of headers\n",
    "maffei.columns = maffei.columns.droplevel([0,1])\n",
    "\n",
    "# Substitute the first 4 headers\n",
    "maffei = maffei.set_axis(header[:4] + maffei.columns[4:].tolist(), axis=1)\n",
    "\n",
    "# Drop last column (who was kept for some mysterious reason)\n",
    "maffei = maffei.iloc[:, :10]\n",
    "\n",
    "# Rename columns\n",
    "change_names = dict()\n",
    "change_names= {'ICTV family (subfamily)': 'family',\n",
    "               'ICTV genus': 'genus',\n",
    "               'Bas#': 'bas',\n",
    "               'Phage (name)': 'phage'}\n",
    "maffei.rename(columns=change_names, inplace=True)\n",
    "del change_names\n",
    "maffei.columns = maffei.columns.str.replace(' ', '_')\n",
    "maffei.columns = maffei.columns.str.replace('.', '')\n",
    "\n",
    "# Move bas column to first position\n",
    "bas = maffei.pop('bas')\n",
    "maffei.insert(0, 'bas', bas)\n",
    "del bas\n",
    "\n",
    "# Define schema\n",
    "header = maffei.columns[:4].tolist()\n",
    "maffei[header] = maffei[header].astype('string')\n",
    "maffei.iloc[:, 4:] = maffei.iloc[:, 4:].astype('int64') # Could be bool or categorical, but int64 is more convenient for machine learning purposes\n",
    "del header\n",
    "\n",
    "display(maffei.head())\n",
    "maffei.dtypes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Basel receptors\n",
    "Basel receptors describes the receptors for each phage in Maffei's paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bas</th>\n",
       "      <th>genus</th>\n",
       "      <th>phage</th>\n",
       "      <th>morphotype</th>\n",
       "      <th>closest_relative</th>\n",
       "      <th>primary_receptor</th>\n",
       "      <th>terminal_receptor</th>\n",
       "      <th>receptor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bas01</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage AugustePiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>RTP (AM156909.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>LptD</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bas02</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage JeanPiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>CEB_EC3a (NC_047812.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>LptD</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bas03</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage JulesPiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>CEB_EC3a (NC_047812.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>FhuA</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bas04</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage FritzSarasin</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>tonnikala (NC_049817.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>BtuB</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bas05</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage PeterMerian</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>XY3 (MN781674.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>FhuA</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bas         genus                             phage  morphotype  \\\n",
       "0  Bas01      Rtpvirus  Escherichia phage AugustePiccard  siphovirus   \n",
       "1  Bas02   Guelphvirus     Escherichia phage JeanPiccard  siphovirus   \n",
       "2  Bas03   Guelphvirus    Escherichia phage JulesPiccard  siphovirus   \n",
       "3  Bas04  Warwickvirus    Escherichia phage FritzSarasin  siphovirus   \n",
       "4  Bas05  Warwickvirus     Escherichia phage PeterMerian  siphovirus   \n",
       "\n",
       "          closest_relative  primary_receptor terminal_receptor receptor  \n",
       "0         RTP (AM156909.1)  LPS / O-antigen?              LptD     n.a.  \n",
       "1   CEB_EC3a (NC_047812.1)  LPS / O-antigen?              LptD     n.a.  \n",
       "2   CEB_EC3a (NC_047812.1)  LPS / O-antigen?              FhuA     n.a.  \n",
       "3  tonnikala (NC_049817.1)  LPS / O-antigen?              BtuB     n.a.  \n",
       "4         XY3 (MN781674.1)  LPS / O-antigen?              FhuA     n.a.  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "bas                  string[python]\n",
       "genus                string[python]\n",
       "phage                string[python]\n",
       "morphotype           string[python]\n",
       "closest_relative     string[python]\n",
       "primary_receptor     string[python]\n",
       "terminal_receptor    string[python]\n",
       "receptor             string[python]\n",
       "dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "receptors = pd.read_csv(os.path.join(RAW_PATH, 'basel_receptors.csv'), sep=';')\n",
    "\n",
    "# Rename columns\n",
    "change_names = {'ICTV genus': 'genus',\n",
    "                'Bas##': 'bas',\n",
    "                'Phage (name)': 'phage',\n",
    "                'closest relative (BLASTN total score)': 'closest relative'} \n",
    "receptors.rename(columns=change_names, inplace=True)\n",
    "del change_names\n",
    "receptors.columns = receptors.columns.str.replace(' ', '_')\n",
    "\n",
    "# Move genus and phage columns to second and third position\n",
    "genus = receptors.pop('genus')\n",
    "phage = receptors.pop('phage')\n",
    "receptors.insert(1, 'genus', genus)\n",
    "receptors.insert(2, 'phage', phage)\n",
    "del genus, phage\n",
    "\n",
    "# Define schema\n",
    "receptors = receptors.astype('string')\n",
    "\n",
    "display(receptors.head())\n",
    "receptors.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Phage protein labels\n",
    "Phage protein is a list of preotein sequences IDs (FASTA) for the viruses in Maffei's paper. It states the categorization of these proteins according to the following Neural Networks:\n",
    "\n",
    "        PhaNNS\n",
    "        PhageRBPdetect\n",
    "        ESM-based method developed by Yumeng\\\n",
    "\\\n",
    "Interpreting the Labels:\n",
    "\n",
    "        Columns 1-11 present the label scores for PhaNNs, with values approaching 10 signifying high confidence.\n",
    "        Column 12 provides PhaNNs' confidence level.\n",
    "        Column 13 indicates PhageRBPdetect predictions: 1 denotes an RBP prediction, while 0 signifies otherwise.\n",
    "        Column 14 offers PhageRBPdetect scores, with values close to 1 signifying strong confidence.\n",
    "        Column 15 presents the ESM-based label.\n",
    "        Column 16 features 1 if the label in Column 15 is \"tail_fiber.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>major_capsid</th>\n",
       "      <th>minor_capsid</th>\n",
       "      <th>baseplate</th>\n",
       "      <th>major_tail</th>\n",
       "      <th>minor_tail</th>\n",
       "      <th>portal</th>\n",
       "      <th>tail_fiber</th>\n",
       "      <th>tail_shaft</th>\n",
       "      <th>collar</th>\n",
       "      <th>HTJ</th>\n",
       "      <th>other</th>\n",
       "      <th>confidence</th>\n",
       "      <th>PhageRBPdetect_prediction</th>\n",
       "      <th>PhageRBDdetect_score</th>\n",
       "      <th>ESM_based_label</th>\n",
       "      <th>ESM_based_fiber_prediction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76132.1_1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.38</td>\n",
       "      <td>5.79</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0</td>\n",
       "      <td>2.403247e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76133.1_2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>3.570606e-08</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76134.1_3</td>\n",
       "      <td>0.35</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.26</td>\n",
       "      <td>1.30</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.73</td>\n",
       "      <td>3.51</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0</td>\n",
       "      <td>1.817806e-06</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76135.1_4</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.03</td>\n",
       "      <td>9.32</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>1.184212e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76136.1_5</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.07</td>\n",
       "      <td>9.28</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>8.151848e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              seqID  major_capsid  minor_capsid  baseplate  \\\n",
       "0  lcl|MZ501051.1_prot_QXV76132.1_1          0.22          0.84       0.18   \n",
       "1  lcl|MZ501051.1_prot_QXV76133.1_2          0.00          0.00       0.00   \n",
       "2  lcl|MZ501051.1_prot_QXV76134.1_3          0.35          1.01       0.71   \n",
       "3  lcl|MZ501051.1_prot_QXV76135.1_4          0.01          0.23       0.02   \n",
       "4  lcl|MZ501051.1_prot_QXV76136.1_5          0.05          0.07       0.06   \n",
       "\n",
       "   major_tail  minor_tail  portal  tail_fiber  tail_shaft  collar   HTJ  \\\n",
       "0        0.21        0.34    0.85        0.49        0.24    0.46  0.38   \n",
       "1        0.00        0.00    0.00        0.00        0.00    0.00  0.00   \n",
       "2        0.26        1.30    0.97        0.68        0.25    0.24  0.73   \n",
       "3        0.01        0.03    0.10        0.09        0.01    0.16  0.03   \n",
       "4        0.05        0.13    0.09        0.03        0.07    0.08  0.07   \n",
       "\n",
       "   other  confidence  PhageRBPdetect_prediction  PhageRBDdetect_score  \\\n",
       "0   5.79        0.98                          0          2.403247e-05   \n",
       "1  10.00        1.00                          0          3.570606e-08   \n",
       "2   3.51        0.97                          0          1.817806e-06   \n",
       "3   9.32        1.00                          0          1.184212e-05   \n",
       "4   9.28        1.00                          0          8.151848e-05   \n",
       "\n",
       "  ESM_based_label  ESM_based_fiber_prediction  \n",
       "0           other                           0  \n",
       "1           other                           0  \n",
       "2           other                           0  \n",
       "3           other                           0  \n",
       "4           other                           0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "seqID                         string[python]\n",
       "major_capsid                         float64\n",
       "minor_capsid                         float64\n",
       "baseplate                            float64\n",
       "major_tail                           float64\n",
       "minor_tail                           float64\n",
       "portal                               float64\n",
       "tail_fiber                           float64\n",
       "tail_shaft                           float64\n",
       "collar                               float64\n",
       "HTJ                                  float64\n",
       "other                                float64\n",
       "confidence                           float64\n",
       "PhageRBPdetect_prediction              int64\n",
       "PhageRBDdetect_score                 float64\n",
       "ESM_based_label               string[python]\n",
       "ESM_based_fiber_prediction             int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteins = pd.read_csv(os.path.join(RAW_PATH, 'phage_proteins_labels.csv'), sep=',')\n",
    "\n",
    "# Rename columns\n",
    "proteins.columns = proteins.columns.str.replace('.', '_')\n",
    "proteins.rename(columns=lambda x: x.lower() if proteins.columns.get_loc(x) < 10 else x, inplace=True)\n",
    "proteins.rename(columns=lambda x: x.lower() if proteins.columns.get_loc(x) in [11, 12] else x, inplace=True)\n",
    "proteins.rename(columns={'name': 'seqID'}, inplace=True)\n",
    "\n",
    "# Define schema\n",
    "proteins['seqID'] = proteins['seqID'].astype('string')\n",
    "proteins.iloc[:, 1:13] = proteins.iloc[:, 1:13].astype('float64')\n",
    "proteins['PhageRBPdetect_prediction'] = proteins['PhageRBPdetect_prediction'].astype('int64')\n",
    "proteins['PhageRBDdetect_score'] = proteins['PhageRBDdetect_score'].astype('float64')\n",
    "proteins['ESM_based_label'] = proteins['ESM_based_label'].astype('string')\n",
    "proteins['ESM_based_fiber_prediction'] = proteins['ESM_based_fiber_prediction'].astype('int64')\n",
    "\n",
    "display(proteins.head())\n",
    "proteins.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Basel proteome\n",
    "basel_proteome.fasta includes the full proteome of each phage mentioned in the Maffei et al. paper. This covers a total of 248 phages that have been sequenced and annotated. Specifically, this includes 69 phages from the Basel collection.\n",
    "\n",
    "Let's load the fasta file in a dataframe, with each tag of the header as a column, as well as the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>protein</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>location</th>\n",
       "      <th>gbkey</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>frame</th>\n",
       "      <th>partial</th>\n",
       "      <th>exception</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76132.1_1</td>\n",
       "      <td>bas01_0001</td>\n",
       "      <td>terminase small subunit</td>\n",
       "      <td>QXV76132.1</td>\n",
       "      <td>1..507</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MSKAALKMGEGNFKALYNKKYGDIAMVAINRKYTPEEVFDFAVRYF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76133.1_2</td>\n",
       "      <td>bas01_0002</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>QXV76133.1</td>\n",
       "      <td>526..621</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MKGFIKLFIWYYLLTSISLCVFMLVVKLWLI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76134.1_3</td>\n",
       "      <td>bas01_0003</td>\n",
       "      <td>terminase large subunit</td>\n",
       "      <td>QXV76134.1</td>\n",
       "      <td>609..2177</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MANLIWEEMTSQEKLAVKAISEHSFEGFLRCWFSITQGERYIPNWH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76135.1_4</td>\n",
       "      <td>bas01_0004</td>\n",
       "      <td>putative homing endonuclease</td>\n",
       "      <td>QXV76135.1</td>\n",
       "      <td>2321..2716</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MVAGSLSGNGYLHIRIGDRRVKNHLIIWEMHNGRIPEGMEIDHINH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76136.1_5</td>\n",
       "      <td>bas01_0005</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>QXV76136.1</td>\n",
       "      <td>2788..3168</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MTKKSKAVYLGNTEGEYYGFTVGNEYDVHNYETEDNFIGTFGDDGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              seqID   locus_tag                       protein  \\\n",
       "0  lcl|MZ501051.1_prot_QXV76132.1_1  bas01_0001       terminase small subunit   \n",
       "1  lcl|MZ501051.1_prot_QXV76133.1_2  bas01_0002          hypothetical protein   \n",
       "2  lcl|MZ501051.1_prot_QXV76134.1_3  bas01_0003       terminase large subunit   \n",
       "3  lcl|MZ501051.1_prot_QXV76135.1_4  bas01_0004  putative homing endonuclease   \n",
       "4  lcl|MZ501051.1_prot_QXV76136.1_5  bas01_0005          hypothetical protein   \n",
       "\n",
       "   protein_id    location gbkey  \\\n",
       "0  QXV76132.1      1..507   CDS   \n",
       "1  QXV76133.1    526..621   CDS   \n",
       "2  QXV76134.1   609..2177   CDS   \n",
       "3  QXV76135.1  2321..2716   CDS   \n",
       "4  QXV76136.1  2788..3168   CDS   \n",
       "\n",
       "                                            sequence gene db_xref frame  \\\n",
       "0  MSKAALKMGEGNFKALYNKKYGDIAMVAINRKYTPEEVFDFAVRYF...  NaN     NaN   NaN   \n",
       "1                    MKGFIKLFIWYYLLTSISLCVFMLVVKLWLI  NaN     NaN   NaN   \n",
       "2  MANLIWEEMTSQEKLAVKAISEHSFEGFLRCWFSITQGERYIPNWH...  NaN     NaN   NaN   \n",
       "3  MVAGSLSGNGYLHIRIGDRRVKNHLIIWEMHNGRIPEGMEIDHINH...  NaN     NaN   NaN   \n",
       "4  MTKKSKAVYLGNTEGEYYGFTVGNEYDVHNYETEDNFIGTFGDDGG...  NaN     NaN   NaN   \n",
       "\n",
       "  partial exception  \n",
       "0     NaN       NaN  \n",
       "1     NaN       NaN  \n",
       "2     NaN       NaN  \n",
       "3     NaN       NaN  \n",
       "4     NaN       NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(35416, 12)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basel_proteome = fasta(os.path.join(RAW_PATH, 'basel_proteome.fasta'), 'phage')\n",
    "display(basel_proteome.df.head())\n",
    "basel_proteome.df.dtypes\n",
    "basel_proteome.df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.5 K12 proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>OS</th>\n",
       "      <th>OX</th>\n",
       "      <th>GN</th>\n",
       "      <th>PE</th>\n",
       "      <th>SV</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp|A5A616|MGTS_ECOLI</td>\n",
       "      <td>Small protein MgtS coli (strain K12)</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>83333</td>\n",
       "      <td>mgtS</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MLGNMNVFMAVLGIILFSGFLAAYFSHKWDD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|O32583|THIS_ECOLI</td>\n",
       "      <td>Sulfur carrier protein ThiS coli (strain K12)</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>83333</td>\n",
       "      <td>thiS</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MQILFNDQAMQCAAGQTVHELLEQLDQRQAGAALAINQQIVPREQW...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|P00350|6PGD_ECOLI</td>\n",
       "      <td>6-phosphogluconate dehydrogenase, decarboxylat...</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>83333</td>\n",
       "      <td>gnd</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>MSKQQIGVVGMAVMGRNLALNIESRGYTVSIFNRSREKTEEVIAEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|P00363|FRDA_ECOLI</td>\n",
       "      <td>Fumarate reductase flavoprotein subunit coli (...</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>83333</td>\n",
       "      <td>frdA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>MQTFQADLAIVGAGGAGLRAAIAAAQANPNAKIALISKVYPMRSHT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|P00370|DHE4_ECOLI</td>\n",
       "      <td>NADP-specific glutamate dehydrogenase coli (st...</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>83333</td>\n",
       "      <td>gdhA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MDQTYSLESFLNHVQKRDPNQTEFAQAVREVMTTLWPFLEQNPKYR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id                                               name  \\\n",
       "0  sp|A5A616|MGTS_ECOLI              Small protein MgtS coli (strain K12)    \n",
       "1  sp|O32583|THIS_ECOLI     Sulfur carrier protein ThiS coli (strain K12)    \n",
       "2  sp|P00350|6PGD_ECOLI  6-phosphogluconate dehydrogenase, decarboxylat...   \n",
       "3  sp|P00363|FRDA_ECOLI  Fumarate reductase flavoprotein subunit coli (...   \n",
       "4  sp|P00370|DHE4_ECOLI  NADP-specific glutamate dehydrogenase coli (st...   \n",
       "\n",
       "            OS     OX    GN PE SV  \\\n",
       "0  Escherichia  83333  mgtS  1  1   \n",
       "1  Escherichia  83333  thiS  1  1   \n",
       "2  Escherichia  83333   gnd  1  2   \n",
       "3  Escherichia  83333  frdA  1  3   \n",
       "4  Escherichia  83333  gdhA  1  1   \n",
       "\n",
       "                                            sequence  \n",
       "0                    MLGNMNVFMAVLGIILFSGFLAAYFSHKWDD  \n",
       "1  MQILFNDQAMQCAAGQTVHELLEQLDQRQAGAALAINQQIVPREQW...  \n",
       "2  MSKQQIGVVGMAVMGRNLALNIESRGYTVSIFNRSREKTEEVIAEN...  \n",
       "3  MQTFQADLAIVGAGGAGLRAAIAAAQANPNAKIALISKVYPMRSHT...  \n",
       "4  MDQTYSLESFLNHVQKRDPNQTEFAQAVREVMTTLWPFLEQNPKYR...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "k12_proteome = fasta(os.path.join(RAW_PATH, 'K12.fasta'), 'k12')\n",
    "display(k12_proteome.df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.6 Save clean datesets in <code>data/interim/clean<code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLEAN_PATH = os.path.join('..', 'data', 'interim', 'clean')\n",
    "\n",
    "if not os.path.exists(CLEAN_PATH):\n",
    "    os.makedirs(CLEAN_PATH)\n",
    "\n",
    "maffei.to_csv(os.path.join(CLEAN_PATH, 'maffei.csv'), index=False)\n",
    "receptors.to_csv(os.path.join(CLEAN_PATH, 'receptors.csv'), index=False)\n",
    "proteins.to_csv(os.path.join(CLEAN_PATH, 'proteins.csv'), index=False)\n",
    "basel_proteome.df.to_csv(os.path.join(CLEAN_PATH, 'basel_proteome.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 Data exploration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some of basel receptors are terminal (aka secondary) and some are primary. In both columns there are both LPS and proteins. We only care about proteins, so it might be a good idea to add a column 'receptor_protein' which selects the proteic one from the two. Also, the 'receptor' column has no proteins, so it might be useless for further analysis. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "3aaba4d7d146b378e05fbb3dda40e2190396d76f525e3e0d3229bbdf23a9518e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
