{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 01 - Data Exploration and labelling\n",
    "\n",
    "In Maffei.xlsx, in the sheet summary statistics, it is reported which phage causes lysis in which E.coli strain. Each row represents a phage, and the columns represent the host. The classification is binary (1=lysis, 0=no-lysis). For the sake of the project, we will consider successful infection only when there is lysis."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# Import user defined libraries\n",
    "import sys\n",
    "sys.path.append(\"../src\")\n",
    "from fastaLoad import fasta"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.Load data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Maffei\n",
    "Let's load data from Maffei. The excel sheet is not clean for import and some rows and columns need to be skipped, as well as some header renamed.\\\n",
    "Also, we only care about the phage identifiers and lysis data, so we'll only keep those columns.\\\n",
    "We obtain a maffei dataset with the first 4 columns identifying the virus: bas, family, genus and phage.\\\n",
    "The following columns state host strains and contain binary values for lysis or not: 1 = lysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bas</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>phage</th>\n",
       "      <th>E_coli_UTI89</th>\n",
       "      <th>E_coli_CFT073</th>\n",
       "      <th>E_coli_55989</th>\n",
       "      <th>S_e_Typhimurium_12023s</th>\n",
       "      <th>S_e_Typhimurium_SL1344</th>\n",
       "      <th>E_coli_B_REL606</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bas01</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage AugustePiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bas02</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage JeanPiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bas03</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage JulesPiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bas04</td>\n",
       "      <td>Drexlerviridae; Tempevirinae</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage FritzSarasin</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bas05</td>\n",
       "      <td>Drexlerviridae; Tempevirinae</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage PeterMerian</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>phage N4</td>\n",
       "      <td>Schitoviridae; Enquatrovirinae</td>\n",
       "      <td>Enquatrovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Bas69</td>\n",
       "      <td>Schitoviridae; Enquatrovirinae</td>\n",
       "      <td>Enquatrovirus</td>\n",
       "      <td>Escherichia phage AlfredRasser</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>lambdavir</td>\n",
       "      <td>Siphoviridae</td>\n",
       "      <td>Lambdavirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>P1vir</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Punavirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>P2vir</td>\n",
       "      <td>Myoviridae; Peduovirinae</td>\n",
       "      <td>Peduovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          bas                          family          genus  \\\n",
       "0       Bas01    Drexlerviridae; Braunvirinae    Guelphvirus   \n",
       "1       Bas02    Drexlerviridae; Braunvirinae       Rtpvirus   \n",
       "2       Bas03    Drexlerviridae; Braunvirinae       Rtpvirus   \n",
       "3       Bas04    Drexlerviridae; Tempevirinae   Warwickvirus   \n",
       "4       Bas05    Drexlerviridae; Tempevirinae   Warwickvirus   \n",
       "..        ...                             ...            ...   \n",
       "73   phage N4  Schitoviridae; Enquatrovirinae  Enquatrovirus   \n",
       "74      Bas69  Schitoviridae; Enquatrovirinae  Enquatrovirus   \n",
       "75  lambdavir                    Siphoviridae    Lambdavirus   \n",
       "76      P1vir                      Myoviridae      Punavirus   \n",
       "77      P2vir        Myoviridae; Peduovirinae     Peduovirus   \n",
       "\n",
       "                               phage  E_coli_UTI89  E_coli_CFT073  \\\n",
       "0   Escherichia phage AugustePiccard             0              0   \n",
       "1      Escherichia phage JeanPiccard             0              0   \n",
       "2     Escherichia phage JulesPiccard             0              0   \n",
       "3     Escherichia phage FritzSarasin             0              1   \n",
       "4      Escherichia phage PeterMerian             0              0   \n",
       "..                               ...           ...            ...   \n",
       "73                              n.a.             1              0   \n",
       "74    Escherichia phage AlfredRasser             1              0   \n",
       "75                              n.a.             0              0   \n",
       "76                              n.a.             0              0   \n",
       "77                              n.a.             0              0   \n",
       "\n",
       "    E_coli_55989  S_e_Typhimurium_12023s  S_e_Typhimurium_SL1344  \\\n",
       "0              0                       0                       0   \n",
       "1              0                       0                       0   \n",
       "2              0                       0                       0   \n",
       "3              0                       0                       0   \n",
       "4              0                       0                       0   \n",
       "..           ...                     ...                     ...   \n",
       "73             1                       0                       0   \n",
       "74             1                       0                       0   \n",
       "75             0                       0                       0   \n",
       "76             0                       0                       0   \n",
       "77             0                       0                       0   \n",
       "\n",
       "    E_coli_B_REL606  \n",
       "0                 1  \n",
       "1                 1  \n",
       "2                 1  \n",
       "3                 1  \n",
       "4                 1  \n",
       "..              ...  \n",
       "73                0  \n",
       "74                0  \n",
       "75                1  \n",
       "76                1  \n",
       "77                1  \n",
       "\n",
       "[78 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "bas                       string[python]\n",
       "family                    string[python]\n",
       "genus                     string[python]\n",
       "phage                     string[python]\n",
       "E_coli_UTI89                       int64\n",
       "E_coli_CFT073                      int64\n",
       "E_coli_55989                       int64\n",
       "S_e_Typhimurium_12023s             int64\n",
       "S_e_Typhimurium_SL1344             int64\n",
       "E_coli_B_REL606                    int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Maffei raw data\n",
    "RAW_PATH = os.path.join('..', 'data', 'raw')\n",
    "\n",
    "maffei = pd.read_excel(os.path.join(RAW_PATH, 'Maffei.xlsx'), sheet_name='summary statistics', skiprows=1, header=[0,1,2])\n",
    "maffei.drop(columns=maffei.columns[0], inplace=True)\n",
    "\n",
    "# Store the first level of the first 4 headers in a list\n",
    "header = [col[0] for col in maffei.columns.tolist()[:4]]\n",
    "header.append('lysis observed (yes=1, no=0)')\n",
    "\n",
    "# Keep only selected headers\n",
    "maffei = maffei.loc[:, header]\n",
    "\n",
    "# Drop odd columns up to 7\n",
    "maffei.drop(columns=maffei.columns[1:9:2], inplace=True)\n",
    "\n",
    "# Remove the first two rows of headers\n",
    "maffei.columns = maffei.columns.droplevel([0,1])\n",
    "\n",
    "# Substitute the first 4 headers\n",
    "maffei = maffei.set_axis(header[:4] + maffei.columns[4:].tolist(), axis=1)\n",
    "\n",
    "# Drop last column (who was kept for some mysterious reason)\n",
    "maffei = maffei.iloc[:, :10]\n",
    "\n",
    "# Rename columns\n",
    "change_names = dict()\n",
    "change_names= {'ICTV family (subfamily)': 'family',\n",
    "               'ICTV genus': 'genus',\n",
    "               'Bas#': 'bas',\n",
    "               'Phage (name)': 'phage'}\n",
    "maffei.rename(columns=change_names, inplace=True)\n",
    "del change_names\n",
    "maffei.columns = maffei.columns.str.replace(' ', '_')\n",
    "maffei.columns = maffei.columns.str.replace('.', '')\n",
    "\n",
    "# Move bas column to first position\n",
    "bas = maffei.pop('bas')\n",
    "maffei.insert(0, 'bas', bas)\n",
    "del bas\n",
    "\n",
    "# Define schema\n",
    "header = maffei.columns[:4].tolist()\n",
    "maffei[header] = maffei[header].astype('string')\n",
    "maffei.iloc[:, 4:] = maffei.iloc[:, 4:].astype('int64') # Could be bool or categorical, but int64 is more convenient for machine learning purposes\n",
    "del header\n",
    "\n",
    "display(maffei)\n",
    "maffei.dtypes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Basel receptors\n",
    "Basel receptors describes the receptors for each phage in Maffei's paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bas</th>\n",
       "      <th>genus</th>\n",
       "      <th>phage</th>\n",
       "      <th>morphotype</th>\n",
       "      <th>closest_relative</th>\n",
       "      <th>primary_receptor</th>\n",
       "      <th>terminal_receptor</th>\n",
       "      <th>receptor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bas01</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage AugustePiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>RTP (AM156909.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>LptD</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bas02</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage JeanPiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>CEB_EC3a (NC_047812.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>LptD</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bas03</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage JulesPiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>CEB_EC3a (NC_047812.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>FhuA</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bas04</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage FritzSarasin</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>tonnikala (NC_049817.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>BtuB</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bas05</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage PeterMerian</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>XY3 (MN781674.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>FhuA</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Teseptimavirus</td>\n",
       "      <td>Escherichia phage T7 (reference)</td>\n",
       "      <td>podovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Enquatrovirus</td>\n",
       "      <td>Escherichia phage N4 (reference)</td>\n",
       "      <td>podovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>ECA</td>\n",
       "      <td>NfrA?</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Lambdavirus</td>\n",
       "      <td>Escherichia phage lambdavir (reference)</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>none</td>\n",
       "      <td>LamB</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Punavirus</td>\n",
       "      <td>Escherichia phage P1vir (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Peduovirus</td>\n",
       "      <td>Escherichia phage P2vir (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      bas           genus                                    phage  \\\n",
       "0   Bas01        Rtpvirus         Escherichia phage AugustePiccard   \n",
       "1   Bas02     Guelphvirus            Escherichia phage JeanPiccard   \n",
       "2   Bas03     Guelphvirus           Escherichia phage JulesPiccard   \n",
       "3   Bas04    Warwickvirus           Escherichia phage FritzSarasin   \n",
       "4   Bas05    Warwickvirus            Escherichia phage PeterMerian   \n",
       "..    ...             ...                                      ...   \n",
       "73   n.a.  Teseptimavirus         Escherichia phage T7 (reference)   \n",
       "74   n.a.   Enquatrovirus         Escherichia phage N4 (reference)   \n",
       "75   n.a.     Lambdavirus  Escherichia phage lambdavir (reference)   \n",
       "76   n.a.       Punavirus      Escherichia phage P1vir (reference)   \n",
       "77   n.a.      Peduovirus      Escherichia phage P2vir (reference)   \n",
       "\n",
       "    morphotype         closest_relative  primary_receptor terminal_receptor  \\\n",
       "0   siphovirus         RTP (AM156909.1)  LPS / O-antigen?              LptD   \n",
       "1   siphovirus   CEB_EC3a (NC_047812.1)  LPS / O-antigen?              LptD   \n",
       "2   siphovirus   CEB_EC3a (NC_047812.1)  LPS / O-antigen?              FhuA   \n",
       "3   siphovirus  tonnikala (NC_049817.1)  LPS / O-antigen?              BtuB   \n",
       "4   siphovirus         XY3 (MN781674.1)  LPS / O-antigen?              FhuA   \n",
       "..         ...                      ...               ...               ...   \n",
       "73   podovirus                     n.a.              n.a.              n.a.   \n",
       "74   podovirus                     n.a.               ECA             NfrA?   \n",
       "75  siphovirus                     n.a.              none              LamB   \n",
       "76    myovirus                     n.a.              n.a.              n.a.   \n",
       "77    myovirus                     n.a.              n.a.              n.a.   \n",
       "\n",
       "     receptor  \n",
       "0        n.a.  \n",
       "1        n.a.  \n",
       "2        n.a.  \n",
       "3        n.a.  \n",
       "4        n.a.  \n",
       "..        ...  \n",
       "73  rough LPS  \n",
       "74       n.a.  \n",
       "75       n.a.  \n",
       "76  rough LPS  \n",
       "77  rough LPS  \n",
       "\n",
       "[78 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "bas                  string[python]\n",
       "genus                string[python]\n",
       "phage                string[python]\n",
       "morphotype           string[python]\n",
       "closest_relative     string[python]\n",
       "primary_receptor     string[python]\n",
       "terminal_receptor    string[python]\n",
       "receptor             string[python]\n",
       "dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "receptors = pd.read_csv(os.path.join(RAW_PATH, 'basel_receptors.csv'), sep=';')\n",
    "\n",
    "# Rename columns\n",
    "change_names = {'ICTV genus': 'genus',\n",
    "                'Bas##': 'bas',\n",
    "                'Phage (name)': 'phage',\n",
    "                'closest relative (BLASTN total score)': 'closest relative'} \n",
    "receptors.rename(columns=change_names, inplace=True)\n",
    "del change_names\n",
    "receptors.columns = receptors.columns.str.replace(' ', '_')\n",
    "\n",
    "# Move genus and phage columns to second and third position\n",
    "genus = receptors.pop('genus')\n",
    "phage = receptors.pop('phage')\n",
    "receptors.insert(1, 'genus', genus)\n",
    "receptors.insert(2, 'phage', phage)\n",
    "del genus, phage\n",
    "\n",
    "# Define schema\n",
    "receptors = receptors.astype('string')\n",
    "\n",
    "display(receptors)\n",
    "receptors.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Phage protein labels\n",
    "Phage protein is a list of preotein sequences IDs (FASTA) for the viruses in Maffei's paper. It states the categorization of these proteins according to the following Neural Networks:\n",
    "\n",
    "        PhaNNS\n",
    "        PhageRBPdetect\n",
    "        ESM-based method developed by Yumeng\\\n",
    "\\\n",
    "Interpreting the Labels:\n",
    "\n",
    "        Columns 1-11 present the label scores for PhaNNs, with values approaching 10 signifying high confidence.\n",
    "        Column 12 provides PhaNNs' confidence level.\n",
    "        Column 13 indicates PhageRBPdetect predictions: 1 denotes an RBP prediction, while 0 signifies otherwise.\n",
    "        Column 14 offers PhageRBPdetect scores, with values close to 1 signifying strong confidence.\n",
    "        Column 15 presents the ESM-based label.\n",
    "        Column 16 features 1 if the label in Column 15 is \"tail_fiber.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>major_capsid</th>\n",
       "      <th>minor_capsid</th>\n",
       "      <th>baseplate</th>\n",
       "      <th>major_tail</th>\n",
       "      <th>minor_tail</th>\n",
       "      <th>portal</th>\n",
       "      <th>tail_fiber</th>\n",
       "      <th>tail_shaft</th>\n",
       "      <th>collar</th>\n",
       "      <th>HTJ</th>\n",
       "      <th>other</th>\n",
       "      <th>confidence</th>\n",
       "      <th>PhageRBPdetect_prediction</th>\n",
       "      <th>PhageRBDdetect_score</th>\n",
       "      <th>ESM_based_label</th>\n",
       "      <th>ESM_based_fiber_prediction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76132.1_1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.38</td>\n",
       "      <td>5.79</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0</td>\n",
       "      <td>2.403247e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76133.1_2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>3.570606e-08</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76134.1_3</td>\n",
       "      <td>0.35</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.26</td>\n",
       "      <td>1.30</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.73</td>\n",
       "      <td>3.51</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0</td>\n",
       "      <td>1.817806e-06</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76135.1_4</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.03</td>\n",
       "      <td>9.32</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>1.184212e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76136.1_5</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.07</td>\n",
       "      <td>9.28</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>8.151848e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35411</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79679.1_75</td>\n",
       "      <td>0.41</td>\n",
       "      <td>1.24</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.43</td>\n",
       "      <td>1.48</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.63</td>\n",
       "      <td>4.22</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0</td>\n",
       "      <td>8.595828e-07</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35412</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79680.1_76</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.10</td>\n",
       "      <td>5.88</td>\n",
       "      <td>3.28</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0</td>\n",
       "      <td>5.343628e-08</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35413</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79681.1_77</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.46</td>\n",
       "      <td>9.25</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.539302e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35414</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79682.1_78</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.94</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>4.384229e-07</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35415</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79683.1_79</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.10</td>\n",
       "      <td>9.72</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>4.812695e-06</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35416 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   seqID  major_capsid  minor_capsid  \\\n",
       "0       lcl|MZ501051.1_prot_QXV76132.1_1          0.22          0.84   \n",
       "1       lcl|MZ501051.1_prot_QXV76133.1_2          0.00          0.00   \n",
       "2       lcl|MZ501051.1_prot_QXV76134.1_3          0.35          1.01   \n",
       "3       lcl|MZ501051.1_prot_QXV76135.1_4          0.01          0.23   \n",
       "4       lcl|MZ501051.1_prot_QXV76136.1_5          0.05          0.07   \n",
       "...                                  ...           ...           ...   \n",
       "35411  lcl|HQ259105.1_prot_AEL79679.1_75          0.41          1.24   \n",
       "35412  lcl|HQ259105.1_prot_AEL79680.1_76          0.02          0.17   \n",
       "35413  lcl|HQ259105.1_prot_AEL79681.1_77          0.01          0.07   \n",
       "35414  lcl|HQ259105.1_prot_AEL79682.1_78          0.00          0.00   \n",
       "35415  lcl|HQ259105.1_prot_AEL79683.1_79          0.00          0.01   \n",
       "\n",
       "       baseplate  major_tail  minor_tail  portal  tail_fiber  tail_shaft  \\\n",
       "0           0.18        0.21        0.34    0.85        0.49        0.24   \n",
       "1           0.00        0.00        0.00    0.00        0.00        0.00   \n",
       "2           0.71        0.26        1.30    0.97        0.68        0.25   \n",
       "3           0.02        0.01        0.03    0.10        0.09        0.01   \n",
       "4           0.06        0.05        0.13    0.09        0.03        0.07   \n",
       "...          ...         ...         ...     ...         ...         ...   \n",
       "35411       0.25        0.09        0.18    0.86        0.43        1.48   \n",
       "35412       0.05        0.11        0.12    0.14        0.09        0.04   \n",
       "35413       0.07        0.00        0.05    0.01        0.06        0.00   \n",
       "35414       0.00        0.00        0.01    0.01        0.02        0.00   \n",
       "35415       0.04        0.01        0.03    0.01        0.05        0.01   \n",
       "\n",
       "       collar   HTJ  other  confidence  PhageRBPdetect_prediction  \\\n",
       "0        0.46  0.38   5.79        0.98                          0   \n",
       "1        0.00  0.00  10.00        1.00                          0   \n",
       "2        0.24  0.73   3.51        0.97                          0   \n",
       "3        0.16  0.03   9.32        1.00                          0   \n",
       "4        0.08  0.07   9.28        1.00                          0   \n",
       "...       ...   ...    ...         ...                        ...   \n",
       "35411    0.21  0.63   4.22        0.97                          0   \n",
       "35412    0.10  5.88   3.28        0.80                          0   \n",
       "35413    0.01  0.46   9.25        1.00                          0   \n",
       "35414    0.02  0.00   9.94        1.00                          0   \n",
       "35415    0.02  0.10   9.72        1.00                          0   \n",
       "\n",
       "       PhageRBDdetect_score ESM_based_label  ESM_based_fiber_prediction  \n",
       "0              2.403247e-05           other                           0  \n",
       "1              3.570606e-08           other                           0  \n",
       "2              1.817806e-06           other                           0  \n",
       "3              1.184212e-05           other                           0  \n",
       "4              8.151848e-05           other                           0  \n",
       "...                     ...             ...                         ...  \n",
       "35411          8.595828e-07           other                           0  \n",
       "35412          5.343628e-08           other                           0  \n",
       "35413          2.539302e-05           other                           0  \n",
       "35414          4.384229e-07           other                           0  \n",
       "35415          4.812695e-06           other                           0  \n",
       "\n",
       "[35416 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "proteins = pd.read_csv(os.path.join(RAW_PATH, 'phage_proteins_labels.csv'), sep=',')\n",
    "\n",
    "# Rename columns\n",
    "proteins.columns = proteins.columns.str.replace('.', '_')\n",
    "proteins.rename(columns=lambda x: x.lower() if proteins.columns.get_loc(x) < 10 else x, inplace=True)\n",
    "proteins.rename(columns=lambda x: x.lower() if proteins.columns.get_loc(x) in [11, 12] else x, inplace=True)\n",
    "proteins.rename(columns={'name': 'seqID'}, inplace=True)\n",
    "\n",
    "# Define schema\n",
    "proteins['seqID'] = proteins['seqID'].astype('string')\n",
    "proteins.iloc[:, 1:13] = proteins.iloc[:, 1:13].astype('float64')\n",
    "proteins['PhageRBPdetect_prediction'] = proteins['PhageRBPdetect_prediction'].astype('int64')\n",
    "proteins['PhageRBDdetect_score'] = proteins['PhageRBDdetect_score'].astype('float64')\n",
    "proteins['ESM_based_label'] = proteins['ESM_based_label'].astype('string')\n",
    "proteins['ESM_based_fiber_prediction'] = proteins['ESM_based_fiber_prediction'].astype('int64')\n",
    "\n",
    "display(proteins)\n",
    "# proteins.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Basel proteome\n",
    "basel_proteome.fasta includes the full proteome of each phage mentioned in the Maffei et al. paper. This covers a total of 248 phages that have been sequenced and annotated. Specifically, this includes 69 phages from the Basel collection.\n",
    "\n",
    "Let's load the fasta file in a dataframe, with each tag of the header as a column, as well as the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>protein</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>location</th>\n",
       "      <th>gbkey</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>frame</th>\n",
       "      <th>partial</th>\n",
       "      <th>exception</th>\n",
       "      <th>bas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76132.1_1</td>\n",
       "      <td>bas01_0001</td>\n",
       "      <td>terminase small subunit</td>\n",
       "      <td>QXV76132.1</td>\n",
       "      <td>1..507</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MSKAALKMGEGNFKALYNKKYGDIAMVAINRKYTPEEVFDFAVRYF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76133.1_2</td>\n",
       "      <td>bas01_0002</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>QXV76133.1</td>\n",
       "      <td>526..621</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MKGFIKLFIWYYLLTSISLCVFMLVVKLWLI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76134.1_3</td>\n",
       "      <td>bas01_0003</td>\n",
       "      <td>terminase large subunit</td>\n",
       "      <td>QXV76134.1</td>\n",
       "      <td>609..2177</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MANLIWEEMTSQEKLAVKAISEHSFEGFLRCWFSITQGERYIPNWH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76135.1_4</td>\n",
       "      <td>bas01_0004</td>\n",
       "      <td>putative homing endonuclease</td>\n",
       "      <td>QXV76135.1</td>\n",
       "      <td>2321..2716</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MVAGSLSGNGYLHIRIGDRRVKNHLIIWEMHNGRIPEGMEIDHINH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76136.1_5</td>\n",
       "      <td>bas01_0005</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>QXV76136.1</td>\n",
       "      <td>2788..3168</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MTKKSKAVYLGNTEGEYYGFTVGNEYDVHNYETEDNFIGTFGDDGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35411</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79679.1_75</td>\n",
       "      <td>gp69</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AEL79679.1</td>\n",
       "      <td>complement(69352..70041)</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MRELTKQQIINALPANFKNSVSQELVDTINNITQDQLVAEAFRENF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gp69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35412</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79680.1_76</td>\n",
       "      <td>gp70</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AEL79680.1</td>\n",
       "      <td>70225..70503</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MRVIDSKDEYSEIDHTKAGFMAGDIILTHDEDTFHELSGMMLLENR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gp70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35413</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79681.1_77</td>\n",
       "      <td>gp71</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AEL79681.1</td>\n",
       "      <td>70705..71103</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MSIYAFDIDITGVLRKQEPYYLGEPEVVSKFTAPSQSVFLPDCEES...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gp71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35414</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79682.1_78</td>\n",
       "      <td>gp71.1</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AEL79682.1</td>\n",
       "      <td>71069..71383</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MKFIMTIPSFVEGESMYTTDFSLEEVMNSNPQQVILVPTLENYHEV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gp71.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35415</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79683.1_79</td>\n",
       "      <td>gp71.2</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AEL79683.1</td>\n",
       "      <td>71387..71698</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MHLYCLISYATTCYLSPYTMDNILAGKGSAATILYEPSEEAEVFLG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gp71.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35416 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   seqID   locus_tag  \\\n",
       "0       lcl|MZ501051.1_prot_QXV76132.1_1  bas01_0001   \n",
       "1       lcl|MZ501051.1_prot_QXV76133.1_2  bas01_0002   \n",
       "2       lcl|MZ501051.1_prot_QXV76134.1_3  bas01_0003   \n",
       "3       lcl|MZ501051.1_prot_QXV76135.1_4  bas01_0004   \n",
       "4       lcl|MZ501051.1_prot_QXV76136.1_5  bas01_0005   \n",
       "...                                  ...         ...   \n",
       "35411  lcl|HQ259105.1_prot_AEL79679.1_75        gp69   \n",
       "35412  lcl|HQ259105.1_prot_AEL79680.1_76        gp70   \n",
       "35413  lcl|HQ259105.1_prot_AEL79681.1_77        gp71   \n",
       "35414  lcl|HQ259105.1_prot_AEL79682.1_78      gp71.1   \n",
       "35415  lcl|HQ259105.1_prot_AEL79683.1_79      gp71.2   \n",
       "\n",
       "                            protein  protein_id                  location  \\\n",
       "0           terminase small subunit  QXV76132.1                    1..507   \n",
       "1              hypothetical protein  QXV76133.1                  526..621   \n",
       "2           terminase large subunit  QXV76134.1                 609..2177   \n",
       "3      putative homing endonuclease  QXV76135.1                2321..2716   \n",
       "4              hypothetical protein  QXV76136.1                2788..3168   \n",
       "...                             ...         ...                       ...   \n",
       "35411          hypothetical protein  AEL79679.1  complement(69352..70041)   \n",
       "35412          hypothetical protein  AEL79680.1              70225..70503   \n",
       "35413          hypothetical protein  AEL79681.1              70705..71103   \n",
       "35414          hypothetical protein  AEL79682.1              71069..71383   \n",
       "35415          hypothetical protein  AEL79683.1              71387..71698   \n",
       "\n",
       "      gbkey                                           sequence gene db_xref  \\\n",
       "0       CDS  MSKAALKMGEGNFKALYNKKYGDIAMVAINRKYTPEEVFDFAVRYF...  NaN     NaN   \n",
       "1       CDS                    MKGFIKLFIWYYLLTSISLCVFMLVVKLWLI  NaN     NaN   \n",
       "2       CDS  MANLIWEEMTSQEKLAVKAISEHSFEGFLRCWFSITQGERYIPNWH...  NaN     NaN   \n",
       "3       CDS  MVAGSLSGNGYLHIRIGDRRVKNHLIIWEMHNGRIPEGMEIDHINH...  NaN     NaN   \n",
       "4       CDS  MTKKSKAVYLGNTEGEYYGFTVGNEYDVHNYETEDNFIGTFGDDGG...  NaN     NaN   \n",
       "...     ...                                                ...  ...     ...   \n",
       "35411   CDS  MRELTKQQIINALPANFKNSVSQELVDTINNITQDQLVAEAFRENF...  NaN     NaN   \n",
       "35412   CDS  MRVIDSKDEYSEIDHTKAGFMAGDIILTHDEDTFHELSGMMLLENR...  NaN     NaN   \n",
       "35413   CDS  MSIYAFDIDITGVLRKQEPYYLGEPEVVSKFTAPSQSVFLPDCEES...  NaN     NaN   \n",
       "35414   CDS  MKFIMTIPSFVEGESMYTTDFSLEEVMNSNPQQVILVPTLENYHEV...  NaN     NaN   \n",
       "35415   CDS  MHLYCLISYATTCYLSPYTMDNILAGKGSAATILYEPSEEAEVFLG...  NaN     NaN   \n",
       "\n",
       "      frame partial exception     bas  \n",
       "0       NaN     NaN       NaN   bas01  \n",
       "1       NaN     NaN       NaN   bas01  \n",
       "2       NaN     NaN       NaN   bas01  \n",
       "3       NaN     NaN       NaN   bas01  \n",
       "4       NaN     NaN       NaN   bas01  \n",
       "...     ...     ...       ...     ...  \n",
       "35411   NaN     NaN       NaN    gp69  \n",
       "35412   NaN     NaN       NaN    gp70  \n",
       "35413   NaN     NaN       NaN    gp71  \n",
       "35414   NaN     NaN       NaN  gp71.1  \n",
       "35415   NaN     NaN       NaN  gp71.2  \n",
       "\n",
       "[35416 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(35416, 13)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# basel_proteome = fasta(os.path.join(RAW_PATH, 'basel_proteome.fasta'), 'phage')\n",
    "# basel_proteome.df['bas'] = basel_proteome.df['locus_tag'].str.split('_', expand=True)[0]\n",
    "# display(basel_proteome.df.loc[basel_proteome.df['protein'] == 'tail fiber', :])\n",
    "display(basel_proteome.df)\n",
    "basel_proteome.df.dtypes\n",
    "basel_proteome.df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.5 K12 proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>OS</th>\n",
       "      <th>OX</th>\n",
       "      <th>GN</th>\n",
       "      <th>PE</th>\n",
       "      <th>SV</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp|A5A616|MGTS_ECOLI</td>\n",
       "      <td>Small protein MgtS</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>mgtS</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MLGNMNVFMAVLGIILFSGFLAAYFSHKWDD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|O32583|THIS_ECOLI</td>\n",
       "      <td>Sulfur carrier protein ThiS</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>thiS</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MQILFNDQAMQCAAGQTVHELLEQLDQRQAGAALAINQQIVPREQW...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|P00350|6PGD_ECOLI</td>\n",
       "      <td>6-phosphogluconate dehydrogenase, decarboxylat...</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>gnd</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>MSKQQIGVVGMAVMGRNLALNIESRGYTVSIFNRSREKTEEVIAEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|P00363|FRDA_ECOLI</td>\n",
       "      <td>Fumarate reductase flavoprotein subunit</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>frdA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>MQTFQADLAIVGAGGAGLRAAIAAAQANPNAKIALISKVYPMRSHT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|P00370|DHE4_ECOLI</td>\n",
       "      <td>NADP-specific glutamate dehydrogenase</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>gdhA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MDQTYSLESFLNHVQKRDPNQTEFAQAVREVMTTLWPFLEQNPKYR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id                                               name  \\\n",
       "0  sp|A5A616|MGTS_ECOLI                                Small protein MgtS    \n",
       "1  sp|O32583|THIS_ECOLI                       Sulfur carrier protein ThiS    \n",
       "2  sp|P00350|6PGD_ECOLI  6-phosphogluconate dehydrogenase, decarboxylat...   \n",
       "3  sp|P00363|FRDA_ECOLI           Fumarate reductase flavoprotein subunit    \n",
       "4  sp|P00370|DHE4_ECOLI             NADP-specific glutamate dehydrogenase    \n",
       "\n",
       "                              OS     OX    GN PE SV  \\\n",
       "0  Escherichia coli (strain K12)  83333  mgtS  1  1   \n",
       "1  Escherichia coli (strain K12)  83333  thiS  1  1   \n",
       "2  Escherichia coli (strain K12)  83333   gnd  1  2   \n",
       "3  Escherichia coli (strain K12)  83333  frdA  1  3   \n",
       "4  Escherichia coli (strain K12)  83333  gdhA  1  1   \n",
       "\n",
       "                                            sequence  \n",
       "0                    MLGNMNVFMAVLGIILFSGFLAAYFSHKWDD  \n",
       "1  MQILFNDQAMQCAAGQTVHELLEQLDQRQAGAALAINQQIVPREQW...  \n",
       "2  MSKQQIGVVGMAVMGRNLALNIESRGYTVSIFNRSREKTEEVIAEN...  \n",
       "3  MQTFQADLAIVGAGGAGLRAAIAAAQANPNAKIALISKVYPMRSHT...  \n",
       "4  MDQTYSLESFLNHVQKRDPNQTEFAQAVREVMTTLWPFLEQNPKYR...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "k12_proteome = fasta(os.path.join(RAW_PATH, 'K12.fasta'), 'k12')\n",
    "display(k12_proteome.df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>OS</th>\n",
       "      <th>OX</th>\n",
       "      <th>GN</th>\n",
       "      <th>PE</th>\n",
       "      <th>SV</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2584</th>\n",
       "      <td>sp|P31600|NFRA_ECOLI</td>\n",
       "      <td>Bacteriophage adsorption protein A</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>nfrA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MKENNLNRVIGWSGLLLTSLLSTSALADNIGTSAEELGLSDYRHFV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id                                 name  \\\n",
       "2584  sp|P31600|NFRA_ECOLI  Bacteriophage adsorption protein A    \n",
       "\n",
       "                                 OS     OX    GN PE SV  \\\n",
       "2584  Escherichia coli (strain K12)  83333  nfrA  1  1   \n",
       "\n",
       "                                               sequence  \n",
       "2584  MKENNLNRVIGWSGLLLTSLLSTSALADNIGTSAEELGLSDYRHFV...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(k12_proteome.df.loc[k12_proteome.df['GN'].str.lower() == 'nfra'.lower()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.6 Save clean datesets in <code>data/interim/clean<code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLEAN_PATH = os.path.join('..', 'data', 'interim', 'clean')\n",
    "\n",
    "if not os.path.exists(CLEAN_PATH):\n",
    "    os.makedirs(CLEAN_PATH)\n",
    "\n",
    "maffei.to_csv(os.path.join(CLEAN_PATH, '0_maffei.csv'), index=False)\n",
    "receptors.to_csv(os.path.join(CLEAN_PATH, '0_receptors.csv'), index=False)\n",
    "proteins.to_csv(os.path.join(CLEAN_PATH, '0_proteins.csv'), index=False)\n",
    "basel_proteome.df.to_csv(os.path.join(CLEAN_PATH, '0_basel_proteome.csv'), index=False)\n",
    "k12_proteome.df.to_csv(os.path.join(CLEAN_PATH, '0_k12_proteome.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 Data exploration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some of basel receptors are terminal (aka secondary) and some are primary. In both columns there are both LPS and proteins. We only care about proteins, so it might be a good idea to add a column 'receptor_protein' which selects the proteic one from the two. Also, the 'receptor' column has no proteins, so it might be useless for further analysis. \\\n",
    "\n",
    "Cleaning to be considered:\n",
    "\n",
    "    1. Some receptor entries have a question mark. For the moment, I will discard them\n",
    "    2. If both primary and terminal receptors are n.a., I will discard the entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pf259\\AppData\\Local\\Temp\\ipykernel_2812\\1070805616.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  prot_receptors['receptor_protein'] = np.where(prot_receptors['primary_receptor'].str.lower().isin(GN_set), prot_receptors['primary_receptor'],\n",
      "C:\\Users\\pf259\\AppData\\Local\\Temp\\ipykernel_2812\\1070805616.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  prot_receptors.drop(columns=['primary_receptor', 'terminal_receptor', 'receptor'], inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bas</th>\n",
       "      <th>genus</th>\n",
       "      <th>phage</th>\n",
       "      <th>morphotype</th>\n",
       "      <th>closest_relative</th>\n",
       "      <th>receptor_protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bas01</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage AugustePiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>RTP (AM156909.1)</td>\n",
       "      <td>LptD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bas02</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage JeanPiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>CEB_EC3a (NC_047812.1)</td>\n",
       "      <td>LptD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bas03</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage JulesPiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>CEB_EC3a (NC_047812.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bas04</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage FritzSarasin</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>tonnikala (NC_049817.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bas05</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage PeterMerian</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>XY3 (MN781674.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bas06</td>\n",
       "      <td>Hanrivervirus</td>\n",
       "      <td>Escherichia phage KarlJaspers</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>herni (NC_049823.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bas07</td>\n",
       "      <td>Hanrivervirus</td>\n",
       "      <td>Escherichia phage JakobBernoulli</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>PGN590 (NC_049830.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Bas08</td>\n",
       "      <td>Tlsvirus</td>\n",
       "      <td>Escherichia phage DanielBernoulli</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>TLS (AY308796.1)</td>\n",
       "      <td>TolC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bas09</td>\n",
       "      <td>Changchunvirus</td>\n",
       "      <td>Escherichia phage PaulSarasin</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>W011D (MK778457.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Bas12</td>\n",
       "      <td>Sertoctavirus</td>\n",
       "      <td>Escherichia phage BrunoManser</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>SRT8 (NC_042043.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Bas13</td>\n",
       "      <td>Tunavirus</td>\n",
       "      <td>Escherichia phage LeonhardEuler</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>phi2013 (MT427400.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Bas14</td>\n",
       "      <td>Dhillonvirus</td>\n",
       "      <td>Escherichia phage TheodorHerzl</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>Sodalis SO-1 (GQ502199.1)</td>\n",
       "      <td>LptD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Bas15</td>\n",
       "      <td>Dhillonvirus</td>\n",
       "      <td>Escherichia phage PaulFeyerabend</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>PTXU06 (MK373789.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Bas16</td>\n",
       "      <td>Dhillonvirus</td>\n",
       "      <td>Escherichia phage GeorgBuechner</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>PTXU06 (MK373789.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Bas17</td>\n",
       "      <td>Dhillonvirus</td>\n",
       "      <td>Escherichia phage KarlBarth</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>PTXU06 (MK373789.1)</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bas18</td>\n",
       "      <td>Dhillonvirus</td>\n",
       "      <td>Escherichia phage Oekolampad</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>slur06 (LN881731.1)</td>\n",
       "      <td>LptD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Bas19</td>\n",
       "      <td>Nonagvirus</td>\n",
       "      <td>Escherichia phage ChristophMerian</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>SP8 (MT682705.1)</td>\n",
       "      <td>LptD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Bas20</td>\n",
       "      <td>Nonagvirus</td>\n",
       "      <td>Escherichia phage FritzHoffmann</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>HdSG1 (MK373795.1)</td>\n",
       "      <td>LptD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Bas21</td>\n",
       "      <td>Seuratvirus</td>\n",
       "      <td>Escherichia phage GottfriedDienst</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>SLUR25 (LT907986.1)</td>\n",
       "      <td>LamB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Bas22</td>\n",
       "      <td>Seuratvirus</td>\n",
       "      <td>Escherichia phage KurtStettler</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>SLUR75 (R025197.1)</td>\n",
       "      <td>LamB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Bas23</td>\n",
       "      <td>Seuratvirus</td>\n",
       "      <td>Escherichia phage WildeMaa</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>SLUR75 (R025197.1)</td>\n",
       "      <td>LamB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Bas24</td>\n",
       "      <td>Seuratvirus</td>\n",
       "      <td>Escherichia phage SeppeHuegi</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>SLUR75 (R025197.1)</td>\n",
       "      <td>LamB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Bas25</td>\n",
       "      <td>Seuratvirus</td>\n",
       "      <td>Escherichia phage VogelGryff</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>Aeromonas DO4 (MH230177.1)</td>\n",
       "      <td>LptD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Bas26</td>\n",
       "      <td>Eseptimavirus</td>\n",
       "      <td>Escherichia phage GreteKellenberger</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>100268_sal2 (KU927497.2)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Bas27</td>\n",
       "      <td>Eseptimavirus</td>\n",
       "      <td>Escherichia phage TrudiGerster</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>SB9 (MK867835.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Bas28</td>\n",
       "      <td>Eseptimavirus</td>\n",
       "      <td>Escherichia phage IrmaTschudi</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>Ace (MT833283.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Bas29</td>\n",
       "      <td>Eseptimavirus</td>\n",
       "      <td>Escherichia phage SuperGirl</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>EASG3 (MK373799.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Bas30</td>\n",
       "      <td>Eseptimavirus</td>\n",
       "      <td>Escherichia phage TrudiRoth</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>100268_sal2 (KU927497.2)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Bas31</td>\n",
       "      <td>Tequintavirus</td>\n",
       "      <td>Escherichia phage DaisyDussoix</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>pork29 (MF431732.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Bas32</td>\n",
       "      <td>Tequintavirus</td>\n",
       "      <td>Escherichia phage IrisVonRoten</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>saus111K (MF431734.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Bas33</td>\n",
       "      <td>Tequintavirus</td>\n",
       "      <td>Escherichia phage HildyBeyeler</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>SSP1 (NC_047881.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Bas34</td>\n",
       "      <td>Tequintavirus</td>\n",
       "      <td>Escherichia phage SelmaRatti</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>NBEco002 (NC_048855.1)</td>\n",
       "      <td>BtuB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Bas35</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage WilhelmHis</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>G29-2 (MK327940.1)</td>\n",
       "      <td>Tsx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Bas36</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage Paracelsus</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>O157 typing phage 7 (KP869105.1)</td>\n",
       "      <td>Tsx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Bas37</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage KarlGJung</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>JK08 (LR877332.1)</td>\n",
       "      <td>Tsx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Bas38</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage AugustSocin</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>OE5505 (MK373785.1)</td>\n",
       "      <td>OmpF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Bas39</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage FriedrichMiescher</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>R5505 (MK373786.1)</td>\n",
       "      <td>Tsx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Bas40</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage FelixPlatter</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>O157 typing phage 7 (KP869105.1)</td>\n",
       "      <td>OmpC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Bas41</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage FriedrichZschokke</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>teqhad (MN895434.1)</td>\n",
       "      <td>OmpC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Bas42</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage AndreasVesalius</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>Sf23 (MF158046.1)</td>\n",
       "      <td>OmpC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Bas43</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage TadeuszReichstein</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>RB18 (MH553563.1)</td>\n",
       "      <td>Tsx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Bas44</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage AdolfPortmann</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>R5505 (MK373786.1)</td>\n",
       "      <td>OmpC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Bas45</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage PaulHMueller</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>R5505 (MK373786.1)</td>\n",
       "      <td>OmpA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Bas46</td>\n",
       "      <td>Mosigvirus</td>\n",
       "      <td>Escherichia phage ChristianSchoenbein</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>PhAPEC2 (KF562341.1)</td>\n",
       "      <td>OmpF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Bas47</td>\n",
       "      <td>Mosigvirus</td>\n",
       "      <td>Escherichia phage AlbertHofmann</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>Shf125875 (KM407600.1)</td>\n",
       "      <td>OmpF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tequintavirus</td>\n",
       "      <td>Escherichia phage T5 (reference)</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>FhuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage T2 (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>FadL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage T4D (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>OmpC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage T6 (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tsx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Lambdavirus</td>\n",
       "      <td>Escherichia phage lambdavir (reference)</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>LamB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      bas           genus                                    phage  \\\n",
       "0   Bas01        Rtpvirus         Escherichia phage AugustePiccard   \n",
       "1   Bas02     Guelphvirus            Escherichia phage JeanPiccard   \n",
       "2   Bas03     Guelphvirus           Escherichia phage JulesPiccard   \n",
       "3   Bas04    Warwickvirus           Escherichia phage FritzSarasin   \n",
       "4   Bas05    Warwickvirus            Escherichia phage PeterMerian   \n",
       "5   Bas06   Hanrivervirus            Escherichia phage KarlJaspers   \n",
       "6   Bas07   Hanrivervirus         Escherichia phage JakobBernoulli   \n",
       "7   Bas08        Tlsvirus        Escherichia phage DanielBernoulli   \n",
       "8   Bas09  Changchunvirus            Escherichia phage PaulSarasin   \n",
       "11  Bas12   Sertoctavirus            Escherichia phage BrunoManser   \n",
       "12  Bas13       Tunavirus          Escherichia phage LeonhardEuler   \n",
       "13  Bas14    Dhillonvirus           Escherichia phage TheodorHerzl   \n",
       "14  Bas15    Dhillonvirus         Escherichia phage PaulFeyerabend   \n",
       "15  Bas16    Dhillonvirus          Escherichia phage GeorgBuechner   \n",
       "16  Bas17    Dhillonvirus              Escherichia phage KarlBarth   \n",
       "17  Bas18    Dhillonvirus             Escherichia phage Oekolampad   \n",
       "18  Bas19      Nonagvirus        Escherichia phage ChristophMerian   \n",
       "19  Bas20      Nonagvirus          Escherichia phage FritzHoffmann   \n",
       "20  Bas21     Seuratvirus        Escherichia phage GottfriedDienst   \n",
       "21  Bas22     Seuratvirus           Escherichia phage KurtStettler   \n",
       "22  Bas23     Seuratvirus               Escherichia phage WildeMaa   \n",
       "23  Bas24     Seuratvirus             Escherichia phage SeppeHuegi   \n",
       "24  Bas25     Seuratvirus             Escherichia phage VogelGryff   \n",
       "25  Bas26   Eseptimavirus      Escherichia phage GreteKellenberger   \n",
       "26  Bas27   Eseptimavirus           Escherichia phage TrudiGerster   \n",
       "27  Bas28   Eseptimavirus            Escherichia phage IrmaTschudi   \n",
       "28  Bas29   Eseptimavirus              Escherichia phage SuperGirl   \n",
       "29  Bas30   Eseptimavirus              Escherichia phage TrudiRoth   \n",
       "30  Bas31   Tequintavirus           Escherichia phage DaisyDussoix   \n",
       "31  Bas32   Tequintavirus           Escherichia phage IrisVonRoten   \n",
       "32  Bas33   Tequintavirus           Escherichia phage HildyBeyeler   \n",
       "33  Bas34   Tequintavirus             Escherichia phage SelmaRatti   \n",
       "34  Bas35   Tequatrovirus             Escherichia phage WilhelmHis   \n",
       "35  Bas36   Tequatrovirus             Escherichia phage Paracelsus   \n",
       "36  Bas37   Tequatrovirus              Escherichia phage KarlGJung   \n",
       "37  Bas38   Tequatrovirus            Escherichia phage AugustSocin   \n",
       "38  Bas39   Tequatrovirus      Escherichia phage FriedrichMiescher   \n",
       "39  Bas40   Tequatrovirus           Escherichia phage FelixPlatter   \n",
       "40  Bas41   Tequatrovirus      Escherichia phage FriedrichZschokke   \n",
       "41  Bas42   Tequatrovirus        Escherichia phage AndreasVesalius   \n",
       "42  Bas43   Tequatrovirus      Escherichia phage TadeuszReichstein   \n",
       "43  Bas44   Tequatrovirus          Escherichia phage AdolfPortmann   \n",
       "44  Bas45   Tequatrovirus           Escherichia phage PaulHMueller   \n",
       "45  Bas46      Mosigvirus    Escherichia phage ChristianSchoenbein   \n",
       "46  Bas47      Mosigvirus          Escherichia phage AlbertHofmann   \n",
       "69   n.a.   Tequintavirus         Escherichia phage T5 (reference)   \n",
       "70   n.a.   Tequatrovirus         Escherichia phage T2 (reference)   \n",
       "71   n.a.   Tequatrovirus        Escherichia phage T4D (reference)   \n",
       "72   n.a.   Tequatrovirus         Escherichia phage T6 (reference)   \n",
       "75   n.a.     Lambdavirus  Escherichia phage lambdavir (reference)   \n",
       "\n",
       "    morphotype                  closest_relative receptor_protein  \n",
       "0   siphovirus                  RTP (AM156909.1)             LptD  \n",
       "1   siphovirus            CEB_EC3a (NC_047812.1)             LptD  \n",
       "2   siphovirus            CEB_EC3a (NC_047812.1)             FhuA  \n",
       "3   siphovirus           tonnikala (NC_049817.1)             BtuB  \n",
       "4   siphovirus                  XY3 (MN781674.1)             FhuA  \n",
       "5   siphovirus               herni (NC_049823.1)             BtuB  \n",
       "6   siphovirus              PGN590 (NC_049830.1)             FhuA  \n",
       "7   siphovirus                  TLS (AY308796.1)             TolC  \n",
       "8   siphovirus                W011D (MK778457.1)             FhuA  \n",
       "11  siphovirus                SRT8 (NC_042043.1)             FhuA  \n",
       "12  siphovirus              phi2013 (MT427400.1)             FhuA  \n",
       "13  siphovirus         Sodalis SO-1 (GQ502199.1)             LptD  \n",
       "14  siphovirus               PTXU06 (MK373789.1)             FhuA  \n",
       "15  siphovirus               PTXU06 (MK373789.1)             FhuA  \n",
       "16  siphovirus               PTXU06 (MK373789.1)             FhuA  \n",
       "17  siphovirus               slur06 (LN881731.1)             LptD  \n",
       "18  siphovirus                  SP8 (MT682705.1)             LptD  \n",
       "19  siphovirus                HdSG1 (MK373795.1)             LptD  \n",
       "20  siphovirus               SLUR25 (LT907986.1)             LamB  \n",
       "21  siphovirus                SLUR75 (R025197.1)             LamB  \n",
       "22  siphovirus                SLUR75 (R025197.1)             LamB  \n",
       "23  siphovirus                SLUR75 (R025197.1)             LamB  \n",
       "24  siphovirus        Aeromonas DO4 (MH230177.1)             LptD  \n",
       "25  siphovirus          100268_sal2 (KU927497.2)             BtuB  \n",
       "26  siphovirus                  SB9 (MK867835.1)             BtuB  \n",
       "27  siphovirus                  Ace (MT833283.1)             BtuB  \n",
       "28  siphovirus                EASG3 (MK373799.1)             BtuB  \n",
       "29  siphovirus          100268_sal2 (KU927497.2)             BtuB  \n",
       "30  siphovirus               pork29 (MF431732.1)             BtuB  \n",
       "31  siphovirus             saus111K (MF431734.1)             BtuB  \n",
       "32  siphovirus                SSP1 (NC_047881.1)             BtuB  \n",
       "33  siphovirus            NBEco002 (NC_048855.1)             BtuB  \n",
       "34    myovirus                G29-2 (MK327940.1)              Tsx  \n",
       "35    myovirus  O157 typing phage 7 (KP869105.1)              Tsx  \n",
       "36    myovirus                 JK08 (LR877332.1)              Tsx  \n",
       "37    myovirus               OE5505 (MK373785.1)             OmpF  \n",
       "38    myovirus                R5505 (MK373786.1)              Tsx  \n",
       "39    myovirus  O157 typing phage 7 (KP869105.1)             OmpC  \n",
       "40    myovirus               teqhad (MN895434.1)             OmpC  \n",
       "41    myovirus                 Sf23 (MF158046.1)             OmpC  \n",
       "42    myovirus                 RB18 (MH553563.1)              Tsx  \n",
       "43    myovirus                R5505 (MK373786.1)             OmpC  \n",
       "44    myovirus                R5505 (MK373786.1)             OmpA  \n",
       "45    myovirus              PhAPEC2 (KF562341.1)             OmpF  \n",
       "46    myovirus            Shf125875 (KM407600.1)             OmpF  \n",
       "69  siphovirus                              n.a.             FhuA  \n",
       "70    myovirus                              n.a.             FadL  \n",
       "71    myovirus                              n.a.             OmpC  \n",
       "72    myovirus                              n.a.              Tsx  \n",
       "75  siphovirus                              n.a.             LamB  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get a set of the elements in a df\n",
    "GN_set = set(k12_proteome.df['GN'].str.lower().tolist())\n",
    "\n",
    "# Create a subset that contains only the rows where either primary_receptor or terminal_receptor are in GN_set\n",
    "prot_receptors = receptors[(receptors['primary_receptor'].str.lower().isin(GN_set)) | (receptors['terminal_receptor'].str.lower().isin(GN_set))]\n",
    "prot_receptors['receptor_protein'] = np.where(prot_receptors['primary_receptor'].str.lower().isin(GN_set), prot_receptors['primary_receptor'], \n",
    "                                      np.where(prot_receptors['terminal_receptor'].str.lower().isin(GN_set), prot_receptors['terminal_receptor'], np.nan))\n",
    "prot_receptors.drop(columns=['primary_receptor', 'terminal_receptor', 'receptor'], inplace=True)\n",
    "\n",
    "display(prot_receptors)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we need to select from each Bas (from receptors), which of its proteins (from basel_proteome) is expected to interact with that kind of receptor. It might be a good idea to define a dictionary to do this. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After selecting the involved protein(s) for each Bas, we'll have to link those proteins with the corresponding protein in the k12_proteome. This pairs will be labeled as interacting. All the others as non-interacting"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "3aaba4d7d146b378e05fbb3dda40e2190396d76f525e3e0d3229bbdf23a9518e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
