{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 01_Ingest_data\n",
    "\n",
    "In Maffei.xlsx, in the sheet summary statistics, it is reported which phage causes lysis in which E.coli strain. Each row represents a phage, and the columns represent the host. The classification is binary (1=lysis, 0=no-lysis). For the sake of the project, we will consider successful infection only when there is lysis."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Import user defined libraries\n",
    "import sys\n",
    "sys.path.append(\"../src\")\n",
    "from fastaLoad import fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Maffei\n",
    "Let's load data from Maffei. The excel sheet is not clean for import and some rows and columns need to be skipped, as well as some header renamed.\\\n",
    "Also, we only care about the phage identifiers and lysis data, so we'll only keep those columns.\\\n",
    "We obtain a maffei dataset with the first 4 columns identifying the virus: bas, family, genus and phage.\\\n",
    "The following columns state host strains and contain binary values for lysis or not: 1 = lysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bas</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>phage</th>\n",
       "      <th>E_coli_UTI89</th>\n",
       "      <th>E_coli_CFT073</th>\n",
       "      <th>E_coli_55989</th>\n",
       "      <th>S_e_Typhimurium_12023s</th>\n",
       "      <th>S_e_Typhimurium_SL1344</th>\n",
       "      <th>E_coli_B_REL606</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bas01</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage AugustePiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bas02</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage JeanPiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bas03</td>\n",
       "      <td>Drexlerviridae; Braunvirinae</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage JulesPiccard</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bas04</td>\n",
       "      <td>Drexlerviridae; Tempevirinae</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage FritzSarasin</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bas05</td>\n",
       "      <td>Drexlerviridae; Tempevirinae</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage PeterMerian</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>phage N4</td>\n",
       "      <td>Schitoviridae; Enquatrovirinae</td>\n",
       "      <td>Enquatrovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Bas69</td>\n",
       "      <td>Schitoviridae; Enquatrovirinae</td>\n",
       "      <td>Enquatrovirus</td>\n",
       "      <td>Escherichia phage AlfredRasser</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>lambdavir</td>\n",
       "      <td>Siphoviridae</td>\n",
       "      <td>Lambdavirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>P1vir</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Punavirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>P2vir</td>\n",
       "      <td>Myoviridae; Peduovirinae</td>\n",
       "      <td>Peduovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          bas                          family          genus  \\\n",
       "0       Bas01    Drexlerviridae; Braunvirinae    Guelphvirus   \n",
       "1       Bas02    Drexlerviridae; Braunvirinae       Rtpvirus   \n",
       "2       Bas03    Drexlerviridae; Braunvirinae       Rtpvirus   \n",
       "3       Bas04    Drexlerviridae; Tempevirinae   Warwickvirus   \n",
       "4       Bas05    Drexlerviridae; Tempevirinae   Warwickvirus   \n",
       "..        ...                             ...            ...   \n",
       "73   phage N4  Schitoviridae; Enquatrovirinae  Enquatrovirus   \n",
       "74      Bas69  Schitoviridae; Enquatrovirinae  Enquatrovirus   \n",
       "75  lambdavir                    Siphoviridae    Lambdavirus   \n",
       "76      P1vir                      Myoviridae      Punavirus   \n",
       "77      P2vir        Myoviridae; Peduovirinae     Peduovirus   \n",
       "\n",
       "                               phage  E_coli_UTI89  E_coli_CFT073  \\\n",
       "0   Escherichia phage AugustePiccard             0              0   \n",
       "1      Escherichia phage JeanPiccard             0              0   \n",
       "2     Escherichia phage JulesPiccard             0              0   \n",
       "3     Escherichia phage FritzSarasin             0              1   \n",
       "4      Escherichia phage PeterMerian             0              0   \n",
       "..                               ...           ...            ...   \n",
       "73                              n.a.             1              0   \n",
       "74    Escherichia phage AlfredRasser             1              0   \n",
       "75                              n.a.             0              0   \n",
       "76                              n.a.             0              0   \n",
       "77                              n.a.             0              0   \n",
       "\n",
       "    E_coli_55989  S_e_Typhimurium_12023s  S_e_Typhimurium_SL1344  \\\n",
       "0              0                       0                       0   \n",
       "1              0                       0                       0   \n",
       "2              0                       0                       0   \n",
       "3              0                       0                       0   \n",
       "4              0                       0                       0   \n",
       "..           ...                     ...                     ...   \n",
       "73             1                       0                       0   \n",
       "74             1                       0                       0   \n",
       "75             0                       0                       0   \n",
       "76             0                       0                       0   \n",
       "77             0                       0                       0   \n",
       "\n",
       "    E_coli_B_REL606  \n",
       "0                 1  \n",
       "1                 1  \n",
       "2                 1  \n",
       "3                 1  \n",
       "4                 1  \n",
       "..              ...  \n",
       "73                0  \n",
       "74                0  \n",
       "75                1  \n",
       "76                1  \n",
       "77                1  \n",
       "\n",
       "[78 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "bas                       string[python]\n",
       "family                    string[python]\n",
       "genus                     string[python]\n",
       "phage                     string[python]\n",
       "E_coli_UTI89                       int64\n",
       "E_coli_CFT073                      int64\n",
       "E_coli_55989                       int64\n",
       "S_e_Typhimurium_12023s             int64\n",
       "S_e_Typhimurium_SL1344             int64\n",
       "E_coli_B_REL606                    int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Maffei raw data\n",
    "RAW_PATH = os.path.join('..', 'data', 'raw')\n",
    "\n",
    "maffei = pd.read_excel(os.path.join(RAW_PATH, 'Maffei.xlsx'), sheet_name='summary statistics', skiprows=1, header=[0,1,2])\n",
    "maffei.drop(columns=maffei.columns[0], inplace=True)\n",
    "\n",
    "# Store the first level of the first 4 headers in a list\n",
    "header = [col[0] for col in maffei.columns.tolist()[:4]]\n",
    "header.append('lysis observed (yes=1, no=0)')\n",
    "\n",
    "# Keep only selected headers\n",
    "maffei = maffei.loc[:, header]\n",
    "\n",
    "# Drop odd columns up to 7\n",
    "maffei.drop(columns=maffei.columns[1:9:2], inplace=True)\n",
    "\n",
    "# Remove the first two rows of headers\n",
    "maffei.columns = maffei.columns.droplevel([0,1])\n",
    "\n",
    "# Substitute the first 4 headers\n",
    "maffei = maffei.set_axis(header[:4] + maffei.columns[4:].tolist(), axis=1)\n",
    "\n",
    "# Drop last column (who was kept for some mysterious reason)\n",
    "maffei = maffei.iloc[:, :10]\n",
    "\n",
    "# Rename columns\n",
    "change_names = dict()\n",
    "change_names= {'ICTV family (subfamily)': 'family',\n",
    "               'ICTV genus': 'genus',\n",
    "               'Bas#': 'bas',\n",
    "               'Phage (name)': 'phage'}\n",
    "maffei.rename(columns=change_names, inplace=True)\n",
    "del change_names\n",
    "maffei.columns = maffei.columns.str.replace(' ', '_')\n",
    "maffei.columns = maffei.columns.str.replace('.', '')\n",
    "\n",
    "# Move bas column to first position\n",
    "bas = maffei.pop('bas')\n",
    "maffei.insert(0, 'bas', bas)\n",
    "del bas\n",
    "\n",
    "# Define schema\n",
    "header = maffei.columns[:4].tolist()\n",
    "maffei[header] = maffei[header].astype('string')\n",
    "maffei.iloc[:, 4:] = maffei.iloc[:, 4:].astype('int64') # Could be bool or categorical, but int64 is more convenient for machine learning purposes\n",
    "del header\n",
    "\n",
    "display(maffei)\n",
    "maffei.dtypes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 78 entries.\\\n",
    "Notice that the last few entries of pahge are n.a."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Basel receptors\n",
    "Basel receptors describes the receptors for each phage in Maffei's paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bas</th>\n",
       "      <th>genus</th>\n",
       "      <th>phage</th>\n",
       "      <th>morphotype</th>\n",
       "      <th>closest_relative</th>\n",
       "      <th>primary_receptor</th>\n",
       "      <th>terminal_receptor</th>\n",
       "      <th>receptor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bas01</td>\n",
       "      <td>Rtpvirus</td>\n",
       "      <td>Escherichia phage AugustePiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>RTP (AM156909.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>LptD</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bas02</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage JeanPiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>CEB_EC3a (NC_047812.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>LptD</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bas03</td>\n",
       "      <td>Guelphvirus</td>\n",
       "      <td>Escherichia phage JulesPiccard</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>CEB_EC3a (NC_047812.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>FhuA</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bas04</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage FritzSarasin</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>tonnikala (NC_049817.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>BtuB</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bas05</td>\n",
       "      <td>Warwickvirus</td>\n",
       "      <td>Escherichia phage PeterMerian</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>XY3 (MN781674.1)</td>\n",
       "      <td>LPS / O-antigen?</td>\n",
       "      <td>FhuA</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Teseptimavirus</td>\n",
       "      <td>Escherichia phage T7 (reference)</td>\n",
       "      <td>podovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Enquatrovirus</td>\n",
       "      <td>Escherichia phage N4 (reference)</td>\n",
       "      <td>podovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>ECA</td>\n",
       "      <td>NfrA?</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Lambdavirus</td>\n",
       "      <td>Escherichia phage lambdavir (reference)</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>none</td>\n",
       "      <td>LamB</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Punavirus</td>\n",
       "      <td>Escherichia phage P1vir (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Peduovirus</td>\n",
       "      <td>Escherichia phage P2vir (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      bas           genus                                    phage  \\\n",
       "0   Bas01        Rtpvirus         Escherichia phage AugustePiccard   \n",
       "1   Bas02     Guelphvirus            Escherichia phage JeanPiccard   \n",
       "2   Bas03     Guelphvirus           Escherichia phage JulesPiccard   \n",
       "3   Bas04    Warwickvirus           Escherichia phage FritzSarasin   \n",
       "4   Bas05    Warwickvirus            Escherichia phage PeterMerian   \n",
       "..    ...             ...                                      ...   \n",
       "73   n.a.  Teseptimavirus         Escherichia phage T7 (reference)   \n",
       "74   n.a.   Enquatrovirus         Escherichia phage N4 (reference)   \n",
       "75   n.a.     Lambdavirus  Escherichia phage lambdavir (reference)   \n",
       "76   n.a.       Punavirus      Escherichia phage P1vir (reference)   \n",
       "77   n.a.      Peduovirus      Escherichia phage P2vir (reference)   \n",
       "\n",
       "    morphotype         closest_relative  primary_receptor terminal_receptor  \\\n",
       "0   siphovirus         RTP (AM156909.1)  LPS / O-antigen?              LptD   \n",
       "1   siphovirus   CEB_EC3a (NC_047812.1)  LPS / O-antigen?              LptD   \n",
       "2   siphovirus   CEB_EC3a (NC_047812.1)  LPS / O-antigen?              FhuA   \n",
       "3   siphovirus  tonnikala (NC_049817.1)  LPS / O-antigen?              BtuB   \n",
       "4   siphovirus         XY3 (MN781674.1)  LPS / O-antigen?              FhuA   \n",
       "..         ...                      ...               ...               ...   \n",
       "73   podovirus                     n.a.              n.a.              n.a.   \n",
       "74   podovirus                     n.a.               ECA             NfrA?   \n",
       "75  siphovirus                     n.a.              none              LamB   \n",
       "76    myovirus                     n.a.              n.a.              n.a.   \n",
       "77    myovirus                     n.a.              n.a.              n.a.   \n",
       "\n",
       "     receptor  \n",
       "0        n.a.  \n",
       "1        n.a.  \n",
       "2        n.a.  \n",
       "3        n.a.  \n",
       "4        n.a.  \n",
       "..        ...  \n",
       "73  rough LPS  \n",
       "74       n.a.  \n",
       "75       n.a.  \n",
       "76  rough LPS  \n",
       "77  rough LPS  \n",
       "\n",
       "[78 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "bas                  string[python]\n",
       "genus                string[python]\n",
       "phage                string[python]\n",
       "morphotype           string[python]\n",
       "closest_relative     string[python]\n",
       "primary_receptor     string[python]\n",
       "terminal_receptor    string[python]\n",
       "receptor             string[python]\n",
       "dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "receptors = pd.read_csv(os.path.join(RAW_PATH, 'basel_receptors.csv'), sep=';')\n",
    "\n",
    "# Rename columns\n",
    "change_names = {'ICTV genus': 'genus',\n",
    "                'Bas##': 'bas',\n",
    "                'Phage (name)': 'phage',\n",
    "                'closest relative (BLASTN total score)': 'closest relative'} \n",
    "receptors.rename(columns=change_names, inplace=True)\n",
    "del change_names\n",
    "receptors.columns = receptors.columns.str.replace(' ', '_')\n",
    "\n",
    "# Move genus and phage columns to second and third position\n",
    "genus = receptors.pop('genus')\n",
    "phage = receptors.pop('phage')\n",
    "receptors.insert(1, 'genus', genus)\n",
    "receptors.insert(2, 'phage', phage)\n",
    "del genus, phage\n",
    "\n",
    "# Define schema\n",
    "receptors = receptors.astype('string')\n",
    "\n",
    "display(receptors)\n",
    "receptors.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bas</th>\n",
       "      <th>genus</th>\n",
       "      <th>phage</th>\n",
       "      <th>morphotype</th>\n",
       "      <th>closest_relative</th>\n",
       "      <th>primary_receptor</th>\n",
       "      <th>terminal_receptor</th>\n",
       "      <th>receptor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tequintavirus</td>\n",
       "      <td>Escherichia phage T5 (reference)</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>LPS / O-antigen</td>\n",
       "      <td>FhuA</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage T2 (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>FadL</td>\n",
       "      <td>LPS (deep core)</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage T4D (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>OmpC</td>\n",
       "      <td>LPS (deep core)</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tequatrovirus</td>\n",
       "      <td>Escherichia phage T6 (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Tsx</td>\n",
       "      <td>LPS (deep core)</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Teseptimavirus</td>\n",
       "      <td>Escherichia phage T7 (reference)</td>\n",
       "      <td>podovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Enquatrovirus</td>\n",
       "      <td>Escherichia phage N4 (reference)</td>\n",
       "      <td>podovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>ECA</td>\n",
       "      <td>NfrA?</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Lambdavirus</td>\n",
       "      <td>Escherichia phage lambdavir (reference)</td>\n",
       "      <td>siphovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>none</td>\n",
       "      <td>LamB</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Punavirus</td>\n",
       "      <td>Escherichia phage P1vir (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>n.a.</td>\n",
       "      <td>Peduovirus</td>\n",
       "      <td>Escherichia phage P2vir (reference)</td>\n",
       "      <td>myovirus</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>rough LPS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bas           genus                                    phage  morphotype  \\\n",
       "69  n.a.   Tequintavirus         Escherichia phage T5 (reference)  siphovirus   \n",
       "70  n.a.   Tequatrovirus         Escherichia phage T2 (reference)    myovirus   \n",
       "71  n.a.   Tequatrovirus        Escherichia phage T4D (reference)    myovirus   \n",
       "72  n.a.   Tequatrovirus         Escherichia phage T6 (reference)    myovirus   \n",
       "73  n.a.  Teseptimavirus         Escherichia phage T7 (reference)   podovirus   \n",
       "74  n.a.   Enquatrovirus         Escherichia phage N4 (reference)   podovirus   \n",
       "75  n.a.     Lambdavirus  Escherichia phage lambdavir (reference)  siphovirus   \n",
       "76  n.a.       Punavirus      Escherichia phage P1vir (reference)    myovirus   \n",
       "77  n.a.      Peduovirus      Escherichia phage P2vir (reference)    myovirus   \n",
       "\n",
       "   closest_relative primary_receptor terminal_receptor   receptor  \n",
       "69             n.a.  LPS / O-antigen              FhuA       n.a.  \n",
       "70             n.a.             FadL   LPS (deep core)       n.a.  \n",
       "71             n.a.             OmpC   LPS (deep core)       n.a.  \n",
       "72             n.a.              Tsx   LPS (deep core)       n.a.  \n",
       "73             n.a.             n.a.              n.a.  rough LPS  \n",
       "74             n.a.              ECA             NfrA?       n.a.  \n",
       "75             n.a.             none              LamB       n.a.  \n",
       "76             n.a.             n.a.              n.a.  rough LPS  \n",
       "77             n.a.             n.a.              n.a.  rough LPS  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display receptor where bas is nan\n",
    "receptors[receptors.bas == 'n.a.']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file has 78 entries, as expected.\\\n",
    "Notice that the last few bas values are n.a.\\\n",
    "We might join the information from maffei and receptors to fill the gaps "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Phage protein labels\n",
    "Phage protein is a list of preotein sequences IDs (FASTA) for the viruses in Maffei's paper. It states the categorization of these proteins according to the following Neural Networks:\n",
    "\n",
    "        PhaNNS\n",
    "        PhageRBPdetect\n",
    "        ESM-based method developed by Yumeng\\\n",
    "\\\n",
    "Interpreting the Labels:\n",
    "\n",
    "        Columns 1-11 present the label scores for PhaNNs, with values approaching 10 signifying high confidence.\n",
    "        Column 12 provides PhaNNs' confidence level.\n",
    "        Column 13 indicates PhageRBPdetect predictions: 1 denotes an RBP prediction, while 0 signifies otherwise.\n",
    "        Column 14 offers PhageRBPdetect scores, with values close to 1 signifying strong confidence.\n",
    "        Column 15 presents the ESM-based label.\n",
    "        Column 16 features 1 if the label in Column 15 is \"tail_fiber.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>major_capsid</th>\n",
       "      <th>minor_capsid</th>\n",
       "      <th>baseplate</th>\n",
       "      <th>major_tail</th>\n",
       "      <th>minor_tail</th>\n",
       "      <th>portal</th>\n",
       "      <th>tail_fiber</th>\n",
       "      <th>tail_shaft</th>\n",
       "      <th>collar</th>\n",
       "      <th>HTJ</th>\n",
       "      <th>other</th>\n",
       "      <th>confidence</th>\n",
       "      <th>PhageRBPdetect_prediction</th>\n",
       "      <th>PhageRBPdetect_score</th>\n",
       "      <th>ESM_based_label</th>\n",
       "      <th>ESM_based_fiber_prediction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76132.1_1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.38</td>\n",
       "      <td>5.79</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0</td>\n",
       "      <td>2.403247e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76133.1_2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>3.570606e-08</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76134.1_3</td>\n",
       "      <td>0.35</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.26</td>\n",
       "      <td>1.30</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.73</td>\n",
       "      <td>3.51</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0</td>\n",
       "      <td>1.817806e-06</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76135.1_4</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.03</td>\n",
       "      <td>9.32</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>1.184212e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76136.1_5</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.07</td>\n",
       "      <td>9.28</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>8.151848e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35411</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79679.1_75</td>\n",
       "      <td>0.41</td>\n",
       "      <td>1.24</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.43</td>\n",
       "      <td>1.48</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.63</td>\n",
       "      <td>4.22</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0</td>\n",
       "      <td>8.595828e-07</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35412</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79680.1_76</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.10</td>\n",
       "      <td>5.88</td>\n",
       "      <td>3.28</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0</td>\n",
       "      <td>5.343628e-08</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35413</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79681.1_77</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.46</td>\n",
       "      <td>9.25</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.539302e-05</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35414</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79682.1_78</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.94</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>4.384229e-07</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35415</th>\n",
       "      <td>lcl|HQ259105.1_prot_AEL79683.1_79</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.10</td>\n",
       "      <td>9.72</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0</td>\n",
       "      <td>4.812695e-06</td>\n",
       "      <td>other</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35416 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   seqID  major_capsid  minor_capsid  \\\n",
       "0       lcl|MZ501051.1_prot_QXV76132.1_1          0.22          0.84   \n",
       "1       lcl|MZ501051.1_prot_QXV76133.1_2          0.00          0.00   \n",
       "2       lcl|MZ501051.1_prot_QXV76134.1_3          0.35          1.01   \n",
       "3       lcl|MZ501051.1_prot_QXV76135.1_4          0.01          0.23   \n",
       "4       lcl|MZ501051.1_prot_QXV76136.1_5          0.05          0.07   \n",
       "...                                  ...           ...           ...   \n",
       "35411  lcl|HQ259105.1_prot_AEL79679.1_75          0.41          1.24   \n",
       "35412  lcl|HQ259105.1_prot_AEL79680.1_76          0.02          0.17   \n",
       "35413  lcl|HQ259105.1_prot_AEL79681.1_77          0.01          0.07   \n",
       "35414  lcl|HQ259105.1_prot_AEL79682.1_78          0.00          0.00   \n",
       "35415  lcl|HQ259105.1_prot_AEL79683.1_79          0.00          0.01   \n",
       "\n",
       "       baseplate  major_tail  minor_tail  portal  tail_fiber  tail_shaft  \\\n",
       "0           0.18        0.21        0.34    0.85        0.49        0.24   \n",
       "1           0.00        0.00        0.00    0.00        0.00        0.00   \n",
       "2           0.71        0.26        1.30    0.97        0.68        0.25   \n",
       "3           0.02        0.01        0.03    0.10        0.09        0.01   \n",
       "4           0.06        0.05        0.13    0.09        0.03        0.07   \n",
       "...          ...         ...         ...     ...         ...         ...   \n",
       "35411       0.25        0.09        0.18    0.86        0.43        1.48   \n",
       "35412       0.05        0.11        0.12    0.14        0.09        0.04   \n",
       "35413       0.07        0.00        0.05    0.01        0.06        0.00   \n",
       "35414       0.00        0.00        0.01    0.01        0.02        0.00   \n",
       "35415       0.04        0.01        0.03    0.01        0.05        0.01   \n",
       "\n",
       "       collar   HTJ  other  confidence  PhageRBPdetect_prediction  \\\n",
       "0        0.46  0.38   5.79        0.98                          0   \n",
       "1        0.00  0.00  10.00        1.00                          0   \n",
       "2        0.24  0.73   3.51        0.97                          0   \n",
       "3        0.16  0.03   9.32        1.00                          0   \n",
       "4        0.08  0.07   9.28        1.00                          0   \n",
       "...       ...   ...    ...         ...                        ...   \n",
       "35411    0.21  0.63   4.22        0.97                          0   \n",
       "35412    0.10  5.88   3.28        0.80                          0   \n",
       "35413    0.01  0.46   9.25        1.00                          0   \n",
       "35414    0.02  0.00   9.94        1.00                          0   \n",
       "35415    0.02  0.10   9.72        1.00                          0   \n",
       "\n",
       "       PhageRBPdetect_score ESM_based_label  ESM_based_fiber_prediction  \n",
       "0              2.403247e-05           other                           0  \n",
       "1              3.570606e-08           other                           0  \n",
       "2              1.817806e-06           other                           0  \n",
       "3              1.184212e-05           other                           0  \n",
       "4              8.151848e-05           other                           0  \n",
       "...                     ...             ...                         ...  \n",
       "35411          8.595828e-07           other                           0  \n",
       "35412          5.343628e-08           other                           0  \n",
       "35413          2.539302e-05           other                           0  \n",
       "35414          4.384229e-07           other                           0  \n",
       "35415          4.812695e-06           other                           0  \n",
       "\n",
       "[35416 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "seqID                         string[python]\n",
       "major_capsid                         float64\n",
       "minor_capsid                         float64\n",
       "baseplate                            float64\n",
       "major_tail                           float64\n",
       "minor_tail                           float64\n",
       "portal                               float64\n",
       "tail_fiber                           float64\n",
       "tail_shaft                           float64\n",
       "collar                               float64\n",
       "HTJ                                  float64\n",
       "other                                float64\n",
       "confidence                           float64\n",
       "PhageRBPdetect_prediction              int64\n",
       "PhageRBPdetect_score                 float64\n",
       "ESM_based_label               string[python]\n",
       "ESM_based_fiber_prediction             int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_NN = pd.read_csv(os.path.join(RAW_PATH, 'phage_proteins_labels.csv'), sep=',')\n",
    "\n",
    "# Rename columns\n",
    "protein_NN.columns = protein_NN.columns.str.replace('.', '_')\n",
    "protein_NN.rename(columns=lambda x: x.lower() if protein_NN.columns.get_loc(x) < 10 else x, inplace=True)\n",
    "protein_NN.rename(columns=lambda x: x.lower() if protein_NN.columns.get_loc(x) in [11, 12] else x, inplace=True)\n",
    "protein_NN.rename(columns={'name': 'seqID'}, inplace=True)\n",
    "protein_NN.rename(columns={'PhageRBDdetect_score': 'PhageRBPdetect_score'}, inplace=True)\n",
    "\n",
    "# Define schema\n",
    "protein_NN['seqID'] = protein_NN['seqID'].astype('string')\n",
    "protein_NN.iloc[:, 1:13] = protein_NN.iloc[:, 1:13].astype('float64')\n",
    "protein_NN['PhageRBPdetect_prediction'] = protein_NN['PhageRBPdetect_prediction'].astype('int64')\n",
    "protein_NN['PhageRBPdetect_score'] = protein_NN['PhageRBPdetect_score'].astype('float64')\n",
    "protein_NN['ESM_based_label'] = protein_NN['ESM_based_label'].str.strip().astype('string')  # Remove leading and trailing whitespaces\n",
    "protein_NN['ESM_based_fiber_prediction'] = protein_NN['ESM_based_fiber_prediction'].astype('int64')\n",
    "\n",
    "# display proteins where PhageRBPdetect_prediction is 1\n",
    "display(protein_NN)\n",
    "protein_NN.shape\n",
    "protein_NN.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Basel proteome\n",
    "basel_proteome.fasta includes the full proteome of each phage mentioned in the Maffei et al. paper. This covers a total of 248 phages that have been sequenced and annotated. Specifically, this includes 69 phages from the Basel collection.\n",
    "\n",
    "Let's load the fasta file in a dataframe, with each tag of the header as a column, as well as the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load fasta file\n",
    "basel_proteome = fasta(os.path.join(RAW_PATH, 'basel_proteome.fasta'), 'phage')\n",
    "\n",
    "# Add bas column to basel_proteome (capitalized)\n",
    "basel_proteome.df['bas'] = basel_proteome.df['locus_tag'].str.split('_', expand=True)[0]\n",
    "basel_proteome.df['bas'] = basel_proteome.df['bas'].str.capitalize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID_phage</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>protein</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>location</th>\n",
       "      <th>gbkey</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>frame</th>\n",
       "      <th>partial</th>\n",
       "      <th>exception</th>\n",
       "      <th>bas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76132.1_1</td>\n",
       "      <td>bas01_0001</td>\n",
       "      <td>terminase small subunit</td>\n",
       "      <td>QXV76132.1</td>\n",
       "      <td>1..507</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MSKAALKMGEGNFKALYNKKYGDIAMVAINRKYTPEEVFDFAVRYF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76133.1_2</td>\n",
       "      <td>bas01_0002</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>QXV76133.1</td>\n",
       "      <td>526..621</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MKGFIKLFIWYYLLTSISLCVFMLVVKLWLI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76134.1_3</td>\n",
       "      <td>bas01_0003</td>\n",
       "      <td>terminase large subunit</td>\n",
       "      <td>QXV76134.1</td>\n",
       "      <td>609..2177</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MANLIWEEMTSQEKLAVKAISEHSFEGFLRCWFSITQGERYIPNWH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76135.1_4</td>\n",
       "      <td>bas01_0004</td>\n",
       "      <td>putative homing endonuclease</td>\n",
       "      <td>QXV76135.1</td>\n",
       "      <td>2321..2716</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MVAGSLSGNGYLHIRIGDRRVKNHLIIWEMHNGRIPEGMEIDHINH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bas01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|MZ501051.1_prot_QXV76136.1_5</td>\n",
       "      <td>bas01_0005</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>QXV76136.1</td>\n",
       "      <td>2788..3168</td>\n",
       "      <td>CDS</td>\n",
       "      <td>MTKKSKAVYLGNTEGEYYGFTVGNEYDVHNYETEDNFIGTFGDDGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bas01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        seqID_phage   locus_tag                       protein  \\\n",
       "0  lcl|MZ501051.1_prot_QXV76132.1_1  bas01_0001       terminase small subunit   \n",
       "1  lcl|MZ501051.1_prot_QXV76133.1_2  bas01_0002          hypothetical protein   \n",
       "2  lcl|MZ501051.1_prot_QXV76134.1_3  bas01_0003       terminase large subunit   \n",
       "3  lcl|MZ501051.1_prot_QXV76135.1_4  bas01_0004  putative homing endonuclease   \n",
       "4  lcl|MZ501051.1_prot_QXV76136.1_5  bas01_0005          hypothetical protein   \n",
       "\n",
       "   protein_id    location gbkey  \\\n",
       "0  QXV76132.1      1..507   CDS   \n",
       "1  QXV76133.1    526..621   CDS   \n",
       "2  QXV76134.1   609..2177   CDS   \n",
       "3  QXV76135.1  2321..2716   CDS   \n",
       "4  QXV76136.1  2788..3168   CDS   \n",
       "\n",
       "                                            sequence gene db_xref frame  \\\n",
       "0  MSKAALKMGEGNFKALYNKKYGDIAMVAINRKYTPEEVFDFAVRYF...  NaN     NaN   NaN   \n",
       "1                    MKGFIKLFIWYYLLTSISLCVFMLVVKLWLI  NaN     NaN   NaN   \n",
       "2  MANLIWEEMTSQEKLAVKAISEHSFEGFLRCWFSITQGERYIPNWH...  NaN     NaN   NaN   \n",
       "3  MVAGSLSGNGYLHIRIGDRRVKNHLIIWEMHNGRIPEGMEIDHINH...  NaN     NaN   NaN   \n",
       "4  MTKKSKAVYLGNTEGEYYGFTVGNEYDVHNYETEDNFIGTFGDDGG...  NaN     NaN   NaN   \n",
       "\n",
       "  partial exception    bas  \n",
       "0     NaN       NaN  Bas01  \n",
       "1     NaN       NaN  Bas01  \n",
       "2     NaN       NaN  Bas01  \n",
       "3     NaN       NaN  Bas01  \n",
       "4     NaN       NaN  Bas01  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(35416, 13)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display(basel_proteome.df.head())\n",
    "basel_proteome.df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID_phage</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>protein</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>location</th>\n",
       "      <th>gbkey</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>frame</th>\n",
       "      <th>partial</th>\n",
       "      <th>exception</th>\n",
       "      <th>bas</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>lcl|AM156909</th>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|AY247822</th>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|EU547803</th>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|EU877232</th>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>140</td>\n",
       "      <td>140</td>\n",
       "      <td>140</td>\n",
       "      <td>140</td>\n",
       "      <td>140</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|HM997020</th>\n",
       "      <td>273</td>\n",
       "      <td>0</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|JF770475</th>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|JN202312</th>\n",
       "      <td>268</td>\n",
       "      <td>0</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>234</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|JN672684</th>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|JX560968</th>\n",
       "      <td>136</td>\n",
       "      <td>0</td>\n",
       "      <td>136</td>\n",
       "      <td>136</td>\n",
       "      <td>136</td>\n",
       "      <td>136</td>\n",
       "      <td>136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KF208315</th>\n",
       "      <td>271</td>\n",
       "      <td>0</td>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KP064094</th>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KP719134</th>\n",
       "      <td>86</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KT184308</th>\n",
       "      <td>235</td>\n",
       "      <td>0</td>\n",
       "      <td>235</td>\n",
       "      <td>235</td>\n",
       "      <td>235</td>\n",
       "      <td>235</td>\n",
       "      <td>235</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KX130960</th>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KX534335</th>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KX534337</th>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KX534341</th>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|KX778611</th>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|LN881725</th>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|LN881731</th>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|LN881733</th>\n",
       "      <td>273</td>\n",
       "      <td>0</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>13</td>\n",
       "      <td>273</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|LN881735</th>\n",
       "      <td>207</td>\n",
       "      <td>0</td>\n",
       "      <td>207</td>\n",
       "      <td>207</td>\n",
       "      <td>207</td>\n",
       "      <td>207</td>\n",
       "      <td>207</td>\n",
       "      <td>11</td>\n",
       "      <td>207</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MF001363</th>\n",
       "      <td>160</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>160</td>\n",
       "      <td>160</td>\n",
       "      <td>160</td>\n",
       "      <td>160</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MG050172</th>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MG488277</th>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MG770228</th>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MG922974</th>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MG963916</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH051914</th>\n",
       "      <td>126</td>\n",
       "      <td>0</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH051915</th>\n",
       "      <td>255</td>\n",
       "      <td>0</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH051916</th>\n",
       "      <td>260</td>\n",
       "      <td>0</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH051918</th>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH230177</th>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH358458</th>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH370375</th>\n",
       "      <td>156</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "      <td>156</td>\n",
       "      <td>156</td>\n",
       "      <td>156</td>\n",
       "      <td>156</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH370379</th>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>123</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH370380</th>\n",
       "      <td>163</td>\n",
       "      <td>0</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>0</td>\n",
       "      <td>139</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH550421</th>\n",
       "      <td>268</td>\n",
       "      <td>0</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH717100</th>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MH751506</th>\n",
       "      <td>266</td>\n",
       "      <td>0</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MK393882</th>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MK672798</th>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MK972707</th>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MK972708</th>\n",
       "      <td>149</td>\n",
       "      <td>0</td>\n",
       "      <td>149</td>\n",
       "      <td>149</td>\n",
       "      <td>149</td>\n",
       "      <td>149</td>\n",
       "      <td>149</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MK972717</th>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MN019128</th>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MN055691</th>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MN090155</th>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MN149903</th>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MN481366</th>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lcl|MT023084</th>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqID_phage  locus_tag  protein  protein_id  location  gbkey  \\\n",
       "0                                                                            \n",
       "lcl|AM156909           75          0       75          75        75     75   \n",
       "lcl|AY247822           50          0       50          50        50     50   \n",
       "lcl|EU547803           55          0       55          55        55     55   \n",
       "lcl|EU877232          140          0      140         140       140    140   \n",
       "lcl|HM997020          273          0      273         273       273    273   \n",
       "lcl|JF770475           74          0       74          74        74     74   \n",
       "lcl|JN202312          268          0      268         268       268    268   \n",
       "lcl|JN672684           83          0       83          83        83     83   \n",
       "lcl|JX560968          136          0      136         136       136    136   \n",
       "lcl|KF208315          271          0      271         271       271    271   \n",
       "lcl|KP064094           91          0       91          91        91     91   \n",
       "lcl|KP719134           86          0       86          86        86     86   \n",
       "lcl|KT184308          235          0      235         235       235    235   \n",
       "lcl|KX130960           74          0       74          74        74     74   \n",
       "lcl|KX534335           62          0       62          62        62     62   \n",
       "lcl|KX534337           93          0       93          93        93     93   \n",
       "lcl|KX534341           63          0       63          63        63     63   \n",
       "lcl|KX778611           48          0       48          48        48     48   \n",
       "lcl|LN881725           90          0       18          90        90     90   \n",
       "lcl|LN881731           58          0       58          58        58     58   \n",
       "lcl|LN881733          273          0      273         273       273    273   \n",
       "lcl|LN881735          207          0      207         207       207    207   \n",
       "lcl|MF001363          160          0      160         160       160    160   \n",
       "lcl|MG050172           60          0       60          60        60     60   \n",
       "lcl|MG488277           51          0       51          51        51     51   \n",
       "lcl|MG770228           74          0       74          74        74     74   \n",
       "lcl|MG922974           97          0       97          97        97     97   \n",
       "lcl|MG963916           10          0       10          10        10     10   \n",
       "lcl|MH051914          126          0      126         126       126    126   \n",
       "lcl|MH051915          255          0      255         255       255    255   \n",
       "lcl|MH051916          260          0      260         260       260    260   \n",
       "lcl|MH051918           87          0       87          87        87     87   \n",
       "lcl|MH230177           67          0       67          67        67     67   \n",
       "lcl|MH358458           78          0       78          78        78     78   \n",
       "lcl|MH370375          156          0      156         156       156    156   \n",
       "lcl|MH370379          147          0      147         147       147    147   \n",
       "lcl|MH370380          163          0      163         163       163    163   \n",
       "lcl|MH550421          268          0      268         268       268    268   \n",
       "lcl|MH717100           45          0       45          45        45     45   \n",
       "lcl|MH751506          266          0      266         266       266    266   \n",
       "lcl|MK393882          158          0      158         158       158    158   \n",
       "lcl|MK672798           92          0       92          92        92     92   \n",
       "lcl|MK972707          147          0      147         147       147    147   \n",
       "lcl|MK972708          149          0      149         149       149    149   \n",
       "lcl|MK972717          147          0      147         147       147    147   \n",
       "lcl|MN019128           67          0       67          67        67     67   \n",
       "lcl|MN055691           65          0       65          65        65     65   \n",
       "lcl|MN090155           85          0       85          85        85     85   \n",
       "lcl|MN149903           47          0       47          47        47     47   \n",
       "lcl|MN481366           45          0       45          45        45     45   \n",
       "lcl|MT023084           80          0       80          80        80     80   \n",
       "\n",
       "              sequence  gene  db_xref  frame  partial  exception  bas  \n",
       "0                                                                      \n",
       "lcl|AM156909        75    75       75      0        0          0    0  \n",
       "lcl|AY247822        50     0        0      0        0          0    0  \n",
       "lcl|EU547803        55    55        0      0        0          3    0  \n",
       "lcl|EU877232       140   140        0      0        0          0    0  \n",
       "lcl|HM997020       273     0        0      0        0          0    0  \n",
       "lcl|JF770475        74     0        0      0        0          0    0  \n",
       "lcl|JN202312       268   234        0      0        0          0    0  \n",
       "lcl|JN672684        83     0        0      1        0          0    0  \n",
       "lcl|JX560968       136     0        0      0        0          1    0  \n",
       "lcl|KF208315       271    72        0      0        0          0    0  \n",
       "lcl|KP064094        91     0        0      0        0          0    0  \n",
       "lcl|KP719134        86     0        0      0        0          0    0  \n",
       "lcl|KT184308       235    47        0      0        0          0    0  \n",
       "lcl|KX130960        74     0        0      0        0          0    0  \n",
       "lcl|KX534335        62     0        0      0        0          0    0  \n",
       "lcl|KX534337        93     0        0      0        0          0    0  \n",
       "lcl|KX534341        63     0        0      0        0          0    0  \n",
       "lcl|KX778611        48     0        0      0        0          0    0  \n",
       "lcl|LN881725        90     0       90      0        0          0    0  \n",
       "lcl|LN881731        58     1       58      0        0          0    0  \n",
       "lcl|LN881733       273    13      273      0        0          0    0  \n",
       "lcl|LN881735       207    11      207      0        0          0    0  \n",
       "lcl|MF001363       160     0        0      0        0          0    0  \n",
       "lcl|MG050172        60     0        0      0        0          0    0  \n",
       "lcl|MG488277        51     0        0      0        0          0    0  \n",
       "lcl|MG770228        74     0        0      0        0          0    0  \n",
       "lcl|MG922974        97     0        0      0        0          0    0  \n",
       "lcl|MG963916        10     0        0      0        0          0    0  \n",
       "lcl|MH051914       126     0        0      0        0          0    0  \n",
       "lcl|MH051915       255     0        0      0        0          0    0  \n",
       "lcl|MH051916       260     0        0      0        0          0    0  \n",
       "lcl|MH051918        87     0        0      0        0          0    0  \n",
       "lcl|MH230177        67     0        0      0        0          0    0  \n",
       "lcl|MH358458        78     0        0      0        0          0    0  \n",
       "lcl|MH370375       156     0      142      0        0          0    0  \n",
       "lcl|MH370379       147     0      123      0        0          0    0  \n",
       "lcl|MH370380       163     0      139      0        0          0    0  \n",
       "lcl|MH550421       268    67        0      0        0          0    0  \n",
       "lcl|MH717100        45     0        0      0        0          0    0  \n",
       "lcl|MH751506       266    64        0      0        0          0    0  \n",
       "lcl|MK393882       158     0        0      0        0          0    0  \n",
       "lcl|MK672798        92     0        0      0        0          0    0  \n",
       "lcl|MK972707       147     0        0      0        0          0    0  \n",
       "lcl|MK972708       149     0        0      0        0          0    0  \n",
       "lcl|MK972717       147     0        0      0        0          0    0  \n",
       "lcl|MN019128        67     0        0      0        0          0    0  \n",
       "lcl|MN055691        65     0        0      0        0          0    0  \n",
       "lcl|MN090155        85     0        0      0        0          0    0  \n",
       "lcl|MN149903        47     0       47      0        0          0    0  \n",
       "lcl|MN481366        45     0        0      0        0          0    0  \n",
       "lcl|MT023084        80     0        0      0        0          0    0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display basel_proteome where bas is nan, grouped by seqID_phage, considering the string before the '.'\n",
    "basel_proteome.df[basel_proteome.df.bas.isna()].groupby(basel_proteome.df.seqID_phage.str.split('.', expand=True)[0]).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. K12 proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID_k12</th>\n",
       "      <th>name</th>\n",
       "      <th>OS</th>\n",
       "      <th>OX</th>\n",
       "      <th>GN</th>\n",
       "      <th>PE</th>\n",
       "      <th>SV</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp|A5A616|MGTS_ECOLI</td>\n",
       "      <td>Small protein MgtS</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>mgtS</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MLGNMNVFMAVLGIILFSGFLAAYFSHKWDD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|O32583|THIS_ECOLI</td>\n",
       "      <td>Sulfur carrier protein ThiS</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>thiS</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MQILFNDQAMQCAAGQTVHELLEQLDQRQAGAALAINQQIVPREQW...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|P00350|6PGD_ECOLI</td>\n",
       "      <td>6-phosphogluconate dehydrogenase, decarboxylat...</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>gnd</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>MSKQQIGVVGMAVMGRNLALNIESRGYTVSIFNRSREKTEEVIAEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|P00363|FRDA_ECOLI</td>\n",
       "      <td>Fumarate reductase flavoprotein subunit</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>frdA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>MQTFQADLAIVGAGGAGLRAAIAAAQANPNAKIALISKVYPMRSHT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|P00370|DHE4_ECOLI</td>\n",
       "      <td>NADP-specific glutamate dehydrogenase</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>gdhA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MDQTYSLESFLNHVQKRDPNQTEFAQAVREVMTTLWPFLEQNPKYR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqID_k12                                               name  \\\n",
       "0  sp|A5A616|MGTS_ECOLI                                Small protein MgtS    \n",
       "1  sp|O32583|THIS_ECOLI                       Sulfur carrier protein ThiS    \n",
       "2  sp|P00350|6PGD_ECOLI  6-phosphogluconate dehydrogenase, decarboxylat...   \n",
       "3  sp|P00363|FRDA_ECOLI           Fumarate reductase flavoprotein subunit    \n",
       "4  sp|P00370|DHE4_ECOLI             NADP-specific glutamate dehydrogenase    \n",
       "\n",
       "                              OS     OX    GN PE SV  \\\n",
       "0  Escherichia coli (strain K12)  83333  mgtS  1  1   \n",
       "1  Escherichia coli (strain K12)  83333  thiS  1  1   \n",
       "2  Escherichia coli (strain K12)  83333   gnd  1  2   \n",
       "3  Escherichia coli (strain K12)  83333  frdA  1  3   \n",
       "4  Escherichia coli (strain K12)  83333  gdhA  1  1   \n",
       "\n",
       "                                            sequence  \n",
       "0                    MLGNMNVFMAVLGIILFSGFLAAYFSHKWDD  \n",
       "1  MQILFNDQAMQCAAGQTVHELLEQLDQRQAGAALAINQQIVPREQW...  \n",
       "2  MSKQQIGVVGMAVMGRNLALNIESRGYTVSIFNRSREKTEEVIAEN...  \n",
       "3  MQTFQADLAIVGAGGAGLRAAIAAAQANPNAKIALISKVYPMRSHT...  \n",
       "4  MDQTYSLESFLNHVQKRDPNQTEFAQAVREVMTTLWPFLEQNPKYR...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(4403, 8)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k12_proteome = fasta(os.path.join(RAW_PATH, 'K12.fasta'), 'k12')\n",
    "display(k12_proteome.df.head())\n",
    "k12_proteome.df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Save clean datesets in <code>data/interim/clean<code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLEAN_PATH = os.path.join('..', 'data', 'interim', 'clean')\n",
    "\n",
    "if not os.path.exists(CLEAN_PATH):\n",
    "    os.makedirs(CLEAN_PATH)\n",
    "\n",
    "# Save dataframes to csv\n",
    "maffei.to_csv(os.path.join(CLEAN_PATH, '1_maffei.csv'), index=False)\n",
    "receptors.to_csv(os.path.join(CLEAN_PATH, '1_receptors.csv'), index=False)\n",
    "protein_NN.to_csv(os.path.join(CLEAN_PATH, '1_proteins.csv'), index=False)\n",
    "basel_proteome.df.to_csv(os.path.join(CLEAN_PATH, '1_basel_proteome.csv'), index=False)\n",
    "k12_proteome.df.to_csv(os.path.join(CLEAN_PATH, '1_k12_proteome.csv'), index=False)\n",
    "\n",
    "# Save dataframes to pickle\n",
    "maffei.to_pickle(os.path.join(CLEAN_PATH, '1_maffei.pkl'))\n",
    "receptors.to_pickle(os.path.join(CLEAN_PATH, '1_receptors.pkl'))\n",
    "protein_NN.to_pickle(os.path.join(CLEAN_PATH, '1_proteins.pkl'))\n",
    "basel_proteome.df.to_pickle(os.path.join(CLEAN_PATH, '1_basel_proteome.pkl'))\n",
    "k12_proteome.df.to_pickle(os.path.join(CLEAN_PATH, '1_k12_proteome.pkl'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "3aaba4d7d146b378e05fbb3dda40e2190396d76f525e3e0d3229bbdf23a9518e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
